<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.554">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Evelyn Metzger">
<meta name="dcterms.date" content="2024-05-01">

<title>Blog - Getting Started with the napari-cosmx plugin</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/cropped-NS_Favicon_512x512-32x32.png" rel="icon" type="image/png">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  


</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../link-to-code.html"> 
<span class="menu-text">Code</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Browse Posts by Topic</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/nanostringtech"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/nanostring-technologies"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
          <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <h1 class="title">Getting Started with the <code>napari-cosmx</code> plugin</h1>
                                <div class="quarto-categories">
                <div class="quarto-category">visualization</div>
                <div class="quarto-category">napari</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Evelyn Metzger </p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              NanoString, a Bruker Company
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/eveilyeverafter" target="_blank"><span class="citation" data-cites="eveilyeverafter">@eveilyeverafter</span></a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">May 1, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a>
  <ul class="collapse">
  <li><a href="#what-is-napari" id="toc-what-is-napari" class="nav-link" data-scroll-target="#what-is-napari"><span class="header-section-number">1.1</span> What is Napari?</a></li>
  <li><a href="#what-is-the-napari-cosmx-plugin" id="toc-what-is-the-napari-cosmx-plugin" class="nav-link" data-scroll-target="#what-is-the-napari-cosmx-plugin"><span class="header-section-number">1.2</span> What is the <code>napari-cosmx</code> plugin?</a></li>
  </ul></li>
  <li><a href="#using-the-napari-cosmx-plugin" id="toc-using-the-napari-cosmx-plugin" class="nav-link" data-scroll-target="#using-the-napari-cosmx-plugin"><span class="header-section-number">2</span> Using the <code>napari-cosmx</code> plugin</a>
  <ul class="collapse">
  <li><a href="#sec-installing" id="toc-sec-installing" class="nav-link" data-scroll-target="#sec-installing"><span class="header-section-number">2.1</span> Installing <code>napari-cosmx</code></a></li>
  <li><a href="#sec-napari-ready-files" id="toc-sec-napari-ready-files" class="nav-link" data-scroll-target="#sec-napari-ready-files"><span class="header-section-number">2.2</span> Napari-ready slide folder</a></li>
  <li><a href="#sec-loading-slides" id="toc-sec-loading-slides" class="nav-link" data-scroll-target="#sec-loading-slides"><span class="header-section-number">2.3</span> Loading a slide into <code>napari-cosmx</code></a></li>
  <li><a href="#sec-stitching" id="toc-sec-stitching" class="nav-link" data-scroll-target="#sec-stitching"><span class="header-section-number">2.4</span> How to create slides from Raw data</a>
  <ul class="collapse">
  <li><a href="#sec-export-raw-data" id="toc-sec-export-raw-data" class="nav-link" data-scroll-target="#sec-export-raw-data"><span class="header-section-number">2.4.1</span> 1. Export Raw Data</a></li>
  <li><a href="#sec-launch-napari-cosmx" id="toc-sec-launch-napari-cosmx" class="nav-link" data-scroll-target="#sec-launch-napari-cosmx"><span class="header-section-number">2.4.2</span> 2. Launch <code>napari-cosmx</code></a></li>
  <li><a href="#sec-stitch-images" id="toc-sec-stitch-images" class="nav-link" data-scroll-target="#sec-stitch-images"><span class="header-section-number">2.4.3</span> 3. Stitching images</a></li>
  </ul></li>
  <li><a href="#sec-adding-metadata" id="toc-sec-adding-metadata" class="nav-link" data-scroll-target="#sec-adding-metadata"><span class="header-section-number">2.5</span> Adding and viewing metadata</a></li>
  </ul></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">3</span> Conclusion</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<div class="cell">
<div class="cell-output-display">
<div id="fig-pancreas" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-pancreas-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/pancreas.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pancreas-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1: Nine Fields of View (FOVs) of a whole-transcriptome Pancreas dataset visualized with the <code>napari-cosmx</code> plugin. DAPI and PanCK are shown in blue and green, respectively. Endocrine cells in the Islets of Langerhans can be identified by their transcript abundance of marker genes (points). Red = <em>GCG</em> (alpha cells), orange = <em>INS</em> (beta cells), cyan = <em>SST</em> (delta cells).
</figcaption>
</figure>
</div>
</div>
</div>
<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>The AtoMx™ Spatial Informatics Portal (SIP) is an excellent solution for analyzing multi-omics spatial data generated on the CosMx™ Spatial Molecular Imager (SMI). It offers seamless integration of data analysis with a data viewer in an intuitive web-based interface.</p>
<p>For bespoke analyses, however, it can be useful to extract SMI image and transcript or protein data from the SIP. Version 1.3 of the SIP introduced new exporting abilities that allow raw data and processed data in a variety of formats (<em>e.g.</em>, Seurat, tiledb, “flat files”) to be downloaded. Still, how do we actually <em>view</em> and interact with the images from these downloads?</p>
<p>This blog post is the first in our series on using our <code>napari-cosmx</code> plugin. Like other items in our <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html" target="_blank">CosMx Analysis Scratch Space</a>, the <code>napari-cosmx</code> plugin is experimental so the usual <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html" target="_blank">caveats</a> and <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/license.html" target="_blank">license</a> applies. Future posts will do deep-dives into <em>using</em> the <code>napari-cosmx</code> plugin. This post will show you how to:</p>
<ul>
<li><a href="#sec-installing" class="quarto-xref">Section&nbsp;2.1</a> Install Napari and the <code>napari-cosmx</code> plugin</li>
<li><a href="#sec-napari-ready-files" class="quarto-xref">Section&nbsp;2.2</a> Describe what our “napari-ready” data format is</li>
<li><a href="#sec-loading-slides" class="quarto-xref">Section&nbsp;2.3</a> How to load a napari-ready slide</li>
<li><a href="#sec-stitching" class="quarto-xref">Section&nbsp;2.4</a> How to create a napari-ready slide from AtoMx raw data export</li>
<li><a href="#sec-adding-metadata" class="quarto-xref">Section&nbsp;2.5</a> Minimum example of how to view cell-level metadata</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p><code>napari-cosmx</code> is actively and continuously under development in the RnD groups at NanoString. We do not (yet) have the source code for <code>napari-cosmx</code> opened-sourced. Please be aware that there may be bugs and that it has not gone through the regular level of quality and testing. Our goal here is to bring the capabilities of napari to CosMx users as fast as possible.</p>
</div>
</div>
<section id="what-is-napari" class="level2" data-number="1.1">
<h2 data-number="1.1" class="anchored" data-anchor-id="what-is-napari"><span class="header-section-number">1.1</span> What is Napari?</h2>
<p>Stepping back for a second. What exacty <em>is</em> Napari? <a href="https://napari.org/" target="_blank">Napari</a> is an <a href="https://github.com/napari/napari" target="_blank">open-source</a> Python project that runs a Qt-based desktop GUI for interactive visualization of scientific images. The application has layers of different types, similar to what you might find in application like Photoshop or Procreate.</p>
</section>
<section id="what-is-the-napari-cosmx-plugin" class="level2" data-number="1.2">
<h2 data-number="1.2" class="anchored" data-anchor-id="what-is-the-napari-cosmx-plugin"><span class="header-section-number">1.2</span> What is the <code>napari-cosmx</code> plugin?</h2>
<p>The napari-cosmx plugin enables viewing of data generated by the CosMx SMI platform in Napari. Tissue morphology layers generated by CosMx SMI are stored as <a href="https://zarr.readthedocs.io/en/stable/" target="_blank">zarr files</a> and displayed as image layers and standard controls such as opacity, gamma, and contrast limits can be use to interact with the tissue. Protein results are also viewed as an image layer in Napari. For CosMx RNA experiments, the detected transcripts are viewed as a points layer in Napari. Cell segmentation results are displayed as an image layer of the cell boundaries. The cell shapes can also be colored by metadata such as cell type. <a href="#fig-example-gif" class="quarto-xref">Figure&nbsp;2</a> shows some examples of these in an animated format.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-example-gif" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-example-gif-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/prostate.gif" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-example-gif-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2: Example animation made with <code>napari-cosmx</code> showing ligand-receptor interactions in a healthy prostate sample. Cell types fill in the cell boundaries. When cells are transparent, one can see more easily see the spatial location of <em>S100A8</em> and <em>S100A9</em> ligand RNA transcripts with the <em>TLR4</em> receptor transcript. While animations are certainly not needed for all purposes, this one highlights that the plugin can color RNA transcripts (points layer), cell-level metadata like cell types (labels layer), and cell boundaries (image layer). It can also visualize IF image layers (not shown here; see <a href="#fig-pancreas" class="quarto-xref">Figure&nbsp;1</a> for example of IF staining).
</figcaption>
</figure>
</div>
</div>
</div>
<p>Beyond the basic interactivity and viewing of napari, other posts in this series will provide examples of tips and tricks as well as more advanced analysis. See the <a href="https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html">series topics</a> for what’s coming ahead!</p>
</section>
</section>
<section id="using-the-napari-cosmx-plugin" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> Using the <code>napari-cosmx</code> plugin</h1>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>While Napari and the <code>napari-cosmx</code> can be installed on many systems, CosMx data can be quite large. A slide with a large number of FOVs may very well exceed the capabilities of a standard laptop.</p>
</div>
</div>
<section id="sec-installing" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="sec-installing"><span class="header-section-number">2.1</span> Installing <code>napari-cosmx</code></h2>
<p>The specific download instructions depend on your operating system (see tabs below) but the general procedure is the same: 1) installing Napari 0.4.17 and 2) installing the <code>napari-cosmx</code> plugin.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The <code>napari-cosmx</code> plugin was developed with Napari 0.4.17. There are some breaking changes that we have noticed if using the plugin with newer version of Napari (<em>e.g.</em>, 0.4.18). At the time of writing this post, please make sure to install version 0.4.17.</p>
</div>
</div>
<div class="tabset-margin-container"></div><div class="panel-tabset">
<ul class="nav nav-tabs" role="tablist"><li class="nav-item" role="presentation"><a class="nav-link active" id="tabset-1-1-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-1" role="tab" aria-controls="tabset-1-1" aria-selected="true">Windows Install</a></li><li class="nav-item" role="presentation"><a class="nav-link" id="tabset-1-2-tab" data-bs-toggle="tab" data-bs-target="#tabset-1-2" role="tab" aria-controls="tabset-1-2" aria-selected="false">MacOS/Unix Install</a></li></ul>
<div class="tab-content">
<div id="tabset-1-1" class="tab-pane active" role="tabpanel" aria-labelledby="tabset-1-1-tab">
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Depending on your browser and security settings you may get warnings when downloading or running some of the links below.</p>
</div>
</div>
<section id="part-1-installing-napari" class="level4" data-number="2.1.0.1">
<h4 data-number="2.1.0.1" class="anchored" data-anchor-id="part-1-installing-napari"><span class="header-section-number">2.1.0.1</span> Part 1: Installing Napari</h4>
<div class="columns">
<div class="column" style="width:45%;">
<p>The Napari project contains platform-specific bundled apps for each release that don’t require you to first install a Python environment. You simply run the installer and a link will be added to your Start Menu as with a typical app installation. The napari-CosMx plugin currently expects Napari 0.4.17.</p>
<p><a href="https://github.com/napari/napari/releases/download/v0.4.17/napari-0.4.17-Windows-x86_64.exe">Click to download the Windows Installer</a></p>
</div><div class="column" style="width:5%;">
<!-- empty column to create gap -->
</div><div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<div id="fig-windows-installer" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-windows-installer-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/windows_installer.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-windows-installer-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3: The Windows Installer.
</figcaption>
</figure>
</div>
</div>
</div>
</div><section id="part-2-installing-napari-cosmx" class="level4" data-number="2.1.0.2">
<h4 data-number="2.1.0.2" class="anchored" data-anchor-id="part-2-installing-napari-cosmx"><span class="header-section-number">2.1.0.2</span> Part 2: Installing <code>napari-cosmx</code></h4>
<ol type="1">
<li>Download the <code>whl</code> file and license in <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/assets/napari-cosmx%20releases" target="_blank">asests/napari-cosmx releases</a>.</li>
<li>Go to the folder where the <code>whl</code> file is in your File Explorer and choose <code>Copy as Path</code>.</li>
<li>Now launch Napari from the Start menu in Windows.</li>
</ol>
<blockquote class="blockquote">
<p>Pro tip: If launching for the first time, the application may take a moment to appear. Avoid launching multiple instances.</p>
</blockquote>
<ol start="3" type="1">
<li>In napari, open up the <code>&gt;_</code> button that is located on the bottom left (see <a href="#fig-windows-plugin-command" class="quarto-xref">Figure&nbsp;4</a> for example).</li>
<li>Type <code>pip install</code> into the console (<em>i.e.</em>, with a single space after the word ‘install’).</li>
<li>Paste the location of the <code>whl</code> file</li>
<li>Press enter to execute</li>
<li>You should receive a message in the console that several packages were successfully installed including the <code>napari-cosmx</code> package.</li>
<li><strong>Close and re-start napari</strong></li>
</ol>
<div class="cell">
<div class="cell-output-display">
<div id="fig-windows-plugin-command" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-windows-plugin-command-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/windows_plugin_command.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-windows-plugin-command-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4: Example showing how to install the <code>napari-cosmx</code> file. Your file name and path will look different. Yellow circle shows the location of the &gt;_ ipython prompt.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</div>
</section>
</div>
<div id="tabset-1-2" class="tab-pane" role="tabpanel" aria-labelledby="tabset-1-2-tab">
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>Depending on your browser and security settings you may get warnings when downloading or running some of the links below.</p>
</div>
</div>
<section id="part-1-installing-napari-1" class="level4" data-number="2.1.0.3">
<h4 data-number="2.1.0.3" class="anchored" data-anchor-id="part-1-installing-napari-1"><span class="header-section-number">2.1.0.3</span> Part 1: Installing Napari</h4>
<p>Visit the Napari <a href="https://github.com/napari/napari/releases/tag/v0.4.17" target="_blank">0.4.17 release page</a>. Scroll all the way to the bottom and expand the section that says “Assets” (<a href="#fig-mac-pkg-download" class="quarto-xref">Figure&nbsp;5</a>). Download the <code>pkg</code> file that is appropriate for your operating system (<em>i.e.</em>, <code>napari-0.4.17-macOS-x86_64.pkg</code> for Mac). Open the downloaded file and install via the instructions on screen (you can accept the defaults). When finished, launch Napari via the Applications folder.</p>
<blockquote class="blockquote">
<p>Pro tip: If launching for the first time, the application may take a moment to appear. Avoid launching multiple instances.</p>
</blockquote>
<div class="cell">
<div class="cell-output-display">
<div id="fig-mac-pkg-download" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-mac-pkg-download-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/mac_install_pkg.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mac-pkg-download-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5: Screenshot of Napari packages. Blue highlighted package is appropriate for Mac. Other operating systems’ packages are also available but untested.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="part-2-installing-napari-cosmx-1" class="level4" data-number="2.1.0.4">
<h4 data-number="2.1.0.4" class="anchored" data-anchor-id="part-2-installing-napari-cosmx-1"><span class="header-section-number">2.1.0.4</span> Part 2: Installing <code>napari-cosmx</code></h4>
<ol type="1">
<li>Download the <code>whl</code> file and license in <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/tree/Main/assets/napari-cosmx%20releases" target="_blank">asests/napari-cosmx releases</a>.</li>
<li>In Napari, open up the <code>&gt;_</code> button that is located on the bottom left (see <a href="#fig-mac-plugin-command" class="quarto-xref">Figure&nbsp;6</a> for example).</li>
<li>Type <code>pip install</code> into the console (<em>i.e.</em>, with a single space after the word ‘install’).</li>
<li>Paste the location of the <code>whl</code> file</li>
<li>Press enter to execute</li>
<li>You should receive a message in the console that several packages were successfully installed including the <code>napari-cosmx</code> package.</li>
<li><strong>Close and re-start napari</strong></li>
</ol>
<div class="cell">
<div class="cell-output-display">
<div id="fig-mac-plugin-command" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-mac-plugin-command-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/mac_plugin_command.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-mac-plugin-command-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6: Example showing how to install the <code>napari-cosmx</code> file. Your file name and path will look different. Yellow circle shows the location of the &gt;_ ipython prompt.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</div>
</div>
</div>
</section>
<section id="sec-napari-ready-files" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="sec-napari-ready-files"><span class="header-section-number">2.2</span> Napari-ready slide folder</h2>
<p>The <code>napari-cosmx</code> plugin expects a slide that has been prepared using the <code>stitching</code> (<a href="#sec-stitching" class="quarto-xref">Section&nbsp;2.4</a>) method within the plugin itself. If you do not already have a napari-ready slide, you can download a simple, single-FOV example that we have created <a href="https://nanostring.box.com/s/vewpiiic0e3xk88kamxi358butsw9isc" target="_blank">here)</a>. This single-FOV dataset was derived from the mouse brain public dataset. The full data download is not needed for our current purposes but those interested in exploring the full data can download it <a href="https://nanostring.com/resources/coronal-hippocampus-and-cortex-raw-data-export/" target="_blank">here</a>.</p>
<p>At the very minimum, the napari-ready slide folder contains two top-level elements:</p>
<ol type="1">
<li><code>images</code> folder. Within the <code>images</code> folder, there are subfolders for each immunofluorescence channel and one folder for the cell boundaries (<code>labels</code>).</li>
<li><code>targets.hdf5</code> file that contains the RNA (or protein) targets.</li>
</ol>
<p>An example can be see below (<a href="#fig-napari-ready-folder" class="quarto-xref">Figure&nbsp;7</a>).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-napari-ready-folder" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-napari-ready-folder-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/napari-ready-format.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-napari-ready-folder-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7: Example layout of a napari-ready folder. In this example, the ‘parent folder’ (<em>i.e.</em>, the folder you would drag and drop into Napari) is named <code>single_fov_napari_example</code> and contains the <code>images</code> folder with subfolders containing various zarr files within and the <code>targets.hdf5</code> file.
</figcaption>
</figure>
</div>
</div>
</div>
<p>If you would like to see the details of the zarr file structure for the <code>images</code> folder in our single-FOV example, expand the code chunk blow.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">$</span> tree <span class="at">-f</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="bu">.</span></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="ex">├──</span> ./images</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> ├── ./images/DNA</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── ./images/DNA/0</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; │&nbsp;&nbsp; └── ./images/DNA/0/0</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; │&nbsp;&nbsp;     └── ./images/DNA/0/0/0</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── ./images/DNA/1</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; │&nbsp;&nbsp; └── ./images/DNA/1/0</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; │&nbsp;&nbsp;     └── ./images/DNA/1/0/0</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; ├── ./images/DNA/2</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; │&nbsp;&nbsp; └── ./images/DNA/2/0</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; │&nbsp;&nbsp;     └── ./images/DNA/2/0/0</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp; └── ./images/DNA/3</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp;     └── ./images/DNA/3/0</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> │&nbsp;&nbsp;         └── ./images/DNA/3/0/0</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span> └── ./images/labels</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     ├── ./images/labels/0</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     │&nbsp;&nbsp; └── ./images/labels/0/0</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     │&nbsp;&nbsp;     └── ./images/labels/0/0/0</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     ├── ./images/labels/1</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     │&nbsp;&nbsp; └── ./images/labels/1/0</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     │&nbsp;&nbsp;     └── ./images/labels/1/0/0</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     ├── ./images/labels/2</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     │&nbsp;&nbsp; └── ./images/labels/2/0</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     │&nbsp;&nbsp;     └── ./images/labels/2/0/0</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>     └── ./images/labels/3</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>         └── ./images/labels/3/0</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a><span class="ex">│&nbsp;&nbsp;</span>             └── ./images/labels/3/0/0</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a><span class="ex">└──</span> ./targets.hdf5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can also create a file named <code>_metadata.csv</code> that can be used for cell-level labeling. For more information on that, please see <a href="#sec-adding-metadata" class="quarto-xref">Section&nbsp;2.5</a>.</p>
</section>
<section id="sec-loading-slides" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="sec-loading-slides"><span class="header-section-number">2.3</span> Loading a slide into <code>napari-cosmx</code></h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>You may have noticed that I put this section <em>before</em> the section on how to create (or ‘stitch’) napari-ready files. This is due to the current implementation of the <code>napari-cosmx</code> code base. Specifically, this is because the widget used for stitching is available once we launch the plugin. So in order to be able to stitch, we need to have <em>some</em> pre-existing slide to load into the plugin. There are advanced ways that we could get around this limitation but that’s an advanced topic for another day.</p>
</div>
</div>
<p>To launch and view CosMx data with the <code>napari-cosmx</code> plugin:</p>
<ol type="1">
<li>Navigate to a napari-ready slide folder. If you need a minimum example, see <a href="#sec-napari-ready-files" class="quarto-xref">Section&nbsp;2.2</a> above.</li>
<li>Open Napari from the Start Menu (Windows) or the Application folder (Mac).</li>
<li>Drag the parent folder of the slide into the the Napari application. If you are using the single-FOV example above, this would be the folder named <code>single_fov_napari_example</code>. Otherwise, the napari-ready folder is whichever folder contains <code>images</code> and <code>targets.hdf5</code> (see <a href="#sec-napari-ready-files" class="quarto-xref">Section&nbsp;2.2</a>).</li>
<li>Napari will ask if you would like to open via the <code>napari-cosmx</code> plugin or another method. Select the <code>napari-cosmx</code> plugin and press okay.</li>
</ol>
</section>
<section id="sec-stitching" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="sec-stitching"><span class="header-section-number">2.4</span> How to create slides from Raw data</h2>
<p>At the time of writing, the process to <em>create</em> Napari ready files follows this framework:</p>
<ol type="1">
<li>Export Raw data from AtoMx SIP (v1.3+) (<a href="#sec-export-raw-data" class="quarto-xref">Section&nbsp;2.4.1</a>)</li>
<li>Launch <code>napari-cosmx</code> within Napari (<a href="#sec-launch-napari-cosmx" class="quarto-xref">Section&nbsp;2.4.2</a>)</li>
<li>Use the stitch widget to create napari-ready slide from raw data (<a href="#sec-stitch-images" class="quarto-xref">Section&nbsp;2.4.3</a>)</li>
</ol>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>One of the main advantages of AtoMx SIP is that the data are stored for you. However, <code>napari-cosmx</code> currently requires the raw data downloaded and stored <strong>locally</strong>. Raw data can be quite large (100s of GBs per slide). It is possible to store the data on a networked drive but we have noticed that stitching performance is slower, depending on your network speed. Storing the data on a high-capacity and fast I/O external hard drive may also be an option.</p>
</div>
</div>
<section id="sec-export-raw-data" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="sec-export-raw-data"><span class="header-section-number">2.4.1</span> 1. Export Raw Data</h3>
<p>In AtoMx SIP, in the <code>Study details</code> panel on the upper left, click <code>Export</code> (<a href="#fig-atomx-export1" class="quarto-xref">Figure&nbsp;8</a>).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-atomx-export1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-atomx-export1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/export-fig1.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-atomx-export1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8: Click <code>Export</code> (available in AtoMx SIP version 1.3+).
</figcaption>
</figure>
</div>
</div>
</div>
<p>Configure your export with the options indicated in <a href="#fig-atomx-export2" class="quarto-xref">Figure&nbsp;9</a> and click <code>EXPORT</code>. If you would like to view metadata (optional). You can go ahead and download the Seurat data now or as a separate step (<em>i.e.</em>, Seurat data is not needed for stitching napari files).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-atomx-export2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-atomx-export2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/export-fig2.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-atomx-export2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9: Example configuration for exporting. For more information on extracting cell-level metadata, see <a href="#sec-adding-metadata" class="quarto-xref">Section&nbsp;2.5</a>.
</figcaption>
</figure>
</div>
</div>
</div>
<p>When the export is ready, download the data. You can do this over the sftp protocol in a variety of application. Here, I’m using <a href="https://cyberduck.io/" target="_blank">Cyberduck</a> but you can use other programs.</p>
<p>In Cyberduck, click <code>Open Connection</code>. In the dropdown menu, select <code>SFTP</code> and enter the URL, username, and your (AtoMx) password. Then click <code>Connect</code>. Example: <a href="#fig-atomx-export3" class="quarto-xref">Figure&nbsp;10</a>.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-atomx-export3" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-atomx-export3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/export-fig3.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-atomx-export3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;10: Example configuration for Cyberduck SFTP.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Once connected, find the relevant folder, right click, and select <code>Download As...</code> (<a href="#fig-atomx-export4" class="quarto-xref">Figure&nbsp;11</a>). Choose the location on your computer to store the data. You may be able to store it on a networked drive but this is currently untested.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-atomx-export4" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-atomx-export4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/export-fig4.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-atomx-export4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;11: Download raw data somewhere on your desktop.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="sec-launch-napari-cosmx" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="sec-launch-napari-cosmx"><span class="header-section-number">2.4.2</span> 2. Launch <code>napari-cosmx</code></h3>
<p>In order to use the stitching widget in the plugin, we must first launch the plugin. Currently, the only way to do that is to load an existing napari-ready folder. This can be any Napari slide (<em>e.g.</em>, a previous study or the single-FOV example described in <a href="#sec-napari-ready-files" class="quarto-xref">Section&nbsp;2.2</a>).</p>
</section>
<section id="sec-stitch-images" class="level3" data-number="2.4.3">
<h3 data-number="2.4.3" class="anchored" data-anchor-id="sec-stitch-images"><span class="header-section-number">2.4.3</span> 3. Stitching images</h3>
<p>Once <code>napari-cosmx</code> is launched, the stitching widget is located on the right-hand panel (<a href="#fig-stitch-widget1" class="quarto-xref">Figure&nbsp;12</a>).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-stitch-widget1" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-stitch-widget1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/stitch-widget1.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stitch-widget1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;12: Example showing a launched <code>napari-cosmx</code> plugin with a possibly-unrelated slide. To stitch, a new slide, Click <code>Browse...</code> in the <code>Stitch Images</code> widget located on the right-hand side of Napari.
</figcaption>
</figure>
</div>
</div>
</div>
<p>We need to tell <code>napari-cosmx</code> where the raw data that we exported are located locally. In the right-hand panel there is a <code>Stitch Images</code> widget. Click <code>Browse...</code> and navigate to the parent folder containing the slide’s raw data.</p>
<p>Your downloaded raw data folder name will be unique to your slide and you can rename it to whatever makes sense for your workflow. For this example, I renamed it <code>raw_data</code> (<a href="#fig-stitch-widget2" class="quarto-xref">Figure&nbsp;13</a>). Click on the raw data folder and select <code>Open</code>. In the <code>Stitch Images</code> widget, you should see the path of the raw data folder printed. If an unexpected format was detected, there will be an error message (<em>e.g.</em>, <a href="#fig-stitch-widget3" class="quarto-xref">Figure&nbsp;14</a>).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-stitch-widget2" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-stitch-widget2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/stitch-widget2.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stitch-widget2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;13: Browsing to the location of the raw data that you want to stitched.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="cell-output-display">
<div id="fig-stitch-widget3" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-stitch-widget3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/stitch-widget3.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stitch-widget3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;14: Stitching widget prints the path to the correctly formatted raw data (left) or provides an error message if not formatted correctly (right). Note that only the correctly formatted location can proceed to the next step (choosing otuput folder).
</figcaption>
</figure>
</div>
</div>
</div>
<p>Next, select the location where you want the plugin to return the napari-ready files. It is recommended not to store it in the same location as the raw data. Here, I’m pointing it to a location adjacent to the raw data that I named <code>stitched_example</code> (<a href="#fig-stitch-widget4" class="quarto-xref">Figure&nbsp;15</a>).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-stitch-widget4" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-stitch-widget4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/stitch-widget4.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stitch-widget4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;15: Select output folder.
</figcaption>
</figure>
</div>
</div>
</div>
<p>Finally, click <code>Stitch</code>. Note: currently there is no refreshing or printing of messages. Please do not click <code>Stitch</code> more than once. You may see Napari become unresponsive, see the “spinning beach ball” (Mac), <em>etc.</em> Depending on the number of FOVs, computer configuration, and analyte type, this can take several minutes. Once complete, you should see messages that resemble that of <a href="#fig-stitch-widget5" class="quarto-xref">Figure&nbsp;16</a>. If you see the last line <code>See output folder for results</code>, you successfully converted the raw data into napari-ready files!</p>
<p>To view the results, simply close napari, reopen it, and drag your newly created results into the application.</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-stitch-widget5" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-stitch-widget5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/stitch-widget5.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-stitch-widget5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;16: Messages from a successfully completed stitching run.
</figcaption>
</figure>
</div>
</div>
</div>
<p>That’s it! To view the newly stitched slide, close Napari, re-open it, and drag the folder into Napari.</p>
</section>
</section>
<section id="sec-adding-metadata" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="sec-adding-metadata"><span class="header-section-number">2.5</span> Adding and viewing metadata</h2>
<p>While there will be dedicated posts that discuss tips and tricks for using the <code>napari-cosmx</code> plugin, here I’ll discuss one of the most powerful uses: viewing cell types.</p>
<p>While not needed for the basic stitching, the Seurat file that is downloaded from AtoMx SIP can contain important cell-level information. For example, if cell typing was performed in AtoMx, each cell will have a label with its cell type.</p>
<p>In this section, I’ll show you the basic principle for converting the meta data within the Seurat object into a csv file that can be understood by the <code>napari-cosmx</code> plugin. Users should have a basic understanding of R in order to use this feature. I’ll also need to switch our example dataset since the minimal single-FOV example dataset was from raw data and not analyzed in AtoMx so we don’t have any cell-level cell type information. Here, the specific column of interest will have the prefix <code>RNA_nbclust</code> and suffix <code>clusters</code>. In the code below, we’ll change that name to simply <code>cell_types</code>. We’ll also need a column named <code>cell_ID</code> in the metadata. We need to write the metadata columns to a file specifically named <code>_metadata.csv</code> and have that file located in the napari-ready folder.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This is R code</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(plyr)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># sem_path will be wherever you downloaded your Seurat object</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>sem_path <span class="ot">&lt;-</span> <span class="st">"/path/to/your/seuratObject.RDS"</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(sem_path)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> sem<span class="sc">@</span>meta.data</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="st">"RNA_nbclust"</span>)) <span class="sc">%&gt;%</span> <span class="fu">select</span>(<span class="fu">ends_with</span>(<span class="st">"clusters"</span>))</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(meta)[<span class="dv">1</span>] <span class="ot">&lt;-</span> <span class="st">'cell_types'</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>meta<span class="sc">$</span>cell_ID <span class="ot">&lt;-</span> <span class="fu">row.names</span>(meta) <span class="co"># adds cell_ID column</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(meta) <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>meta <span class="ot">&lt;-</span> meta <span class="sc">%&gt;%</span> <span class="fu">relocate</span>(cell_ID) <span class="co"># moves cell_ID to first column position</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(meta, <span class="at">file=</span><span class="st">"/path/to/inside/napari-ready-folder/_metadata.csv"</span>, </span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>            <span class="at">sep=</span><span class="st">","</span>, <span class="at">col.names=</span><span class="cn">TRUE</span>, <span class="at">row.names=</span><span class="cn">FALSE</span>, <span class="at">quote=</span><span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Now, when we drag and drop the napari-ready folder, the metadata that you extracted from Seurat will be available to view using the right-hand widget named <code>Color Cells</code> (<a href="#fig-color-cells" class="quarto-xref">Figure&nbsp;17</a>).</p>
<div class="cell">
<div class="cell-output-display">
<div id="fig-color-cells" class="quarto-figure quarto-figure-center quarto-float anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-color-cells-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="./figures/color-cells.png" class="img-fluid figure-img">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-color-cells-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;17: When metadata are available in the _metadata.csv file, it’s possible to color cells based on a cell-level metadata value (<em>e.g.</em>, cell types).
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="conclusion" class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Conclusion</h1>
<p>In this post we introduced the <code>napari-cosmx</code> plugin, a tool that can be used to interact with exported CosMx SMI data from AtoMx SIP. Our focus was on installation and launching the application. In future posts we’ll do deep-dives on usage and share tips and tricks that we’ve learned along the way. If you would like to see a Napari topic discussed, please create an issue or feature request on <a href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space/labels/napari%20series" target="_blank">github</a> using the <code>Napari series</code> label.</p>
<p>–ERM</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../license.html">
<p>License</p>
</a>
  </li>  
    <li class="nav-item">
    <a class="nav-link" href="../../how-to-contribute.html">
<p>How to contribute</p>
</a>
  </li>  
</ul>
    <div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>