---
title: "Using Napari to view and analyze CosMx data"
author:
  - name: Evelyn Metzger
    affiliations: 
      - ref: nstg
      - ref: eveilyeverafter
affiliations:
  - id: nstg
    name: NanoString Technologies, Inc.
    city: Seattle
    state: WA
  - id: eveilyeverafter
    name: "Github: [@eveilyeverafter](https://github.com/eveilyeverafter)"
toc: true
toc-title: Contents
toc-depth: 3
toc-expand: 2
toc-location: left
number-sections: true
number-depth: 4
date: "2024-04-19"
categories: [visualization, napari]
draft: false
image: figures/pancreas.png
---

```{r}
#| eval: true
#| echo: false
#| label: "fig-pancreas"
#| fig-cap: "Nine Fields of View (FOVs) of the whole-transcriptome Pancreas dataset generated with the `napari-cosmx` plugin. DAPI and PanCK is shown in blue and green, respectively. Endocrine cells in the Islets of Langerhans can be idenified by their clusters of cells that express marker gene transcripts red (_GCG_; alpha cells), blue (_to do_; to do), and orange (_INS_; beta cells)."

knitr::include_graphics("./figures/pancreas.png")
```

# Introduction 

The AtoMx Spatial Informatics Portal (SIP) is an excellent solution for analyzing 
multi-omics spatial data generated on the CosMx Spatial Molecular Imager (SMI).
It offers seamless integration of data analysis with data viewer in an intuitive 
web-based interface. 

For bespoke analyses, however, it can be useful to extract SMI image and transcript or 
protein data from the SIP. Version 1.3 of the SIP introduced new exporting abilities that allow
raw data and processed data in a variety of formats (_e.g._, Seurat, tiledb, "flat files) 
to be downloaded. Still, how do we actually _view_ and interact with the images from these downloads?

This blog post is the first in our series on using our
`napari-cosmx` plugin. Like other items in our [CosMx Analysis Scratch Space](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html),
the `napari-cosmx` plugin is experimental so the usual [caveats]() and [license](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/license.html) applies. 

::: {.callout-note}
`napari-cosmx` is actively and continuously under development in the RnD groups at
NanoString. We do not (yet) have the source code for `napari-cosmx` opened-sourced.
Please be aware that there may be bugs and that it has not gone through the regular
level of quality and testing. Our goal here is to bring the capabilities of napari
to CosMx users as fast as possible.
:::

## What is Napari?

Stepping back for a second. What exacty _is_ Napari. [Napari](https://napari.org/)
is an [open-source](https://github.com/napari/napari) Python project that runs a 
Qt-based desktop GUI for interactive visualization of scientific images. 
The application has layers of different types, similar to what you might find
in application like Photoshop or Procreate. 

## What is the `napari-cosmx` plugin?

Tissue morphology layers generated by CosMx SMI are displayed as image layers 
in napari, and standard controls are available such as opacity, gamma, and contrast
limits. Protein results are also viewed as an image layer in napari. For CosMx 
RNA experiments, the detected transcripts are viewed as a points layer in napari.
Cell segmentation results are displayed as an image layer of the cell boundaries.
The cell shapes can also be colored by metadata such as cell type.

Beyond the basic interactivity and viewing of napari, other posts in this series
will provide examples of tips and tricks as well as more advanced analysis. See the 
[series topics](https://nanostring-biostats.github.io/CosMx-Analysis-Scratch-Space/about.html) for what's coming ahead!

> **to do** add some figures showing a few of the capabilities of the plugin. 


# Using the `napari-cosmx` plugin 


::: {.callout-note}
While napari and the `napari-cosmx` can be installed on many systems, CosMx data
can be quite large. A slide with a large number of FOVs may very well exceed the
capabilities of a standard laptop.
:::


## Installing `napari-cosmx`

The specific download instructions depend on your operating system (see tabs below)
but the general procedure is the same: 1) installing napari 0.4.17 and 2) installing the `napari-cosmx` plugin

::: {.callout-note}
The `napari-cosmx` plugin was developed with napari 0.4.17. There are some breaking
changes that we have noticed if using the plugin with newer version of napari (_e.g.,
0.4.18). Please make sure 

:::

::: {.panel-tabset}

### Windows Install

::: {.callout-note}
Depending on your browser and security settings you may get warnings when downloading or running some of the links below.
:::

#### Part 1: Installing Napari

:::: {.columns}

::: {.column width="45%"}

The napari project contains platform-specific bundled apps for each release that
donâ€™t require you to first install a Python environment. You simply run the installer
and a link will be added to your Start Menu as with a typical app installation. 
The napari-CosMx plugin currently expects napari 0.4.17.

[Click to download the Windows Installer](https://github.com/napari/napari/releases/download/v0.4.17/napari-0.4.17-Windows-x86_64.exe)

:::

::: {.column width="5%"}
<!-- empty column to create gap -->
:::

::: {.column width="40%"}

```{r}
#| eval: true
#| echo: false
#| label: "fig-windows-installer"
#| fig-cap: "The Windows Installer."

knitr::include_graphics("./figures/windows_installer.png")
```

:::

#### Part 2: Installing `napari-cosmx`

1. Download the whl file (**need a location for this**). 
2. Go to the folder where the `whl` file is in your File Explorer and choose 
`Copy Path`.
2. Now launch napari from the Start menu in Windows. 

> Pro tip: If launching for the first time, the application may take a moment to appear. Avoid launching multiple instances. 

3. In napari, open up the `>_` button that is located on the bottom left (see @fig-windows-plugin-command for example). 
4. Type `pip install ` into the console (_i.e._, with a single space after the word 'install').
5. Paste the location of the `whl` file
6. Press enter to execute
7. You should receive a message in the console that several packages were successfully
installed including the `napari-cosmx` package. 
8. **Close and re-start napari**

```{r}
#| eval: true
#| echo: false
#| label: "fig-windows-plugin-command"
#| fig-cap: "Example showing how to install the `napari-cosmx` file. Your file name and path will look different."

knitr::include_graphics("./figures/windows_plugin_command.png")
```


::::




### MacOS/Unix Install

::: {.callout-note}
Depending on your browser and security settings you may get warnings when downloading or running some of the links below.
:::

#### Part 1: Installing Napari

Visit the napari [0.4.17 release page](https://github.com/napari/napari/releases/tag/v0.4.17){target="_blank"}. 
Scroll all the way to the bottom and expand the section that says "Assets" (@fig-mac-pkg-download). 
Download the `pkg` file that is appropriate for your operating system (_i.e._, 
`napari-0.4.17-macOS-x86_64.pkg` for Mac). Open the downloaded file and install 
via the instructions on screen (you can accept the defaults). When finished, 
launch napari via the Applications folder. 

> Pro tip: If launching for the first time, the application may take a moment to appear. Avoid launching multiple instances. 

```{r}
#| eval: true
#| echo: false
#| label: "fig-mac-pkg-download"
#| fig-cap: "Screenshot of napari packages. Blue highlighted package is appropriate for Mac. Other operating systems' packages are also available but untested."

knitr::include_graphics("./figures/mac_install_pkg.png")
```



#### Part 2: Installing `napari-cosmx`

1. In napari, open up the `>_` button that is located on the bottom left (see @fig-mac-plugin-command for example). 
2. Type `pip install ` into the console (_i.e._, with a single space after the word 'install').
3. Paste the location of the `whl` file
4. Press enter to execute
5. You should receive a message in the console that several packages were successfully
installed including the `napari-cosmx` package. 
6. **Close and re-start napari**

```{r}
#| eval: true
#| echo: false
#| label: "fig-mac-plugin-command"
#| fig-cap: "Example showing how to install the `napari-cosmx` file. Your file name and path will look different."

knitr::include_graphics("./figures/mac_plugin_command.png")
```

:::


## Napari-ready slide folder {#sec-napari-ready-files}

The `napari-cosmx` plugin expects a single slide that has been prepared using the
`stitching` (@sec-stitching) method within the plugin itself. If you do not already
have a napari-ready slide, you can download a two-FOV example that we have created
[here](){target="_blank"}. These two FOVs are derived from the coronal hippocampus and
cortex public dataset. The full data download is not needed for this study but
those interested in the analyzing the full dataset can see it  [here](https://nanostring.com/resources/coronal-hippocampus-and-cortex-raw-data-export/){target="_blank"}). 

At the very minimum, this slide folder contains two elements: 

1. `images` folder. Within the `images` folder, there are subfolders for each Immunofluorescence channel and one folder for the cell boundaries (`labels`). 
2. `targets.hdf5` file that contains the RNA (or protein) targets. 

**to do** make figure showing the layout. 

We can also create a file named `_metadata.csv` that can be used for cell-level
labeling. For more information on that, please see @sec-adding-metadata. 

## Loading a slide into `napari-cosmx` {#sec-loading-slides}

To launch and view CosMx data with the `napari-cosmx`: 

1. Navigate to a napari-ready slide folder. If you need a minimum example, see @sec-napari-ready-files above.
2. Open napari from the Start Menu (Windows) or the Application folder (Mac).
3. Drag the parent directory of the slide into the the napari application. If you
are using the two-fov example above, this would be the folder named `two_fovs`. 
Otherwise, the napari-ready folder is whichever folder containing `images` and `targets.hdf5`.
4. napari will ask if you would like to open via the `napari-cosmx` plugin or another
method. Select the `napari-cosmx` plugin and press okay. 


## How to create slides from Raw data {#sec-stitching}

At the time of writing, the process to _create_ napari ready files follows this framework: 

1. Export Raw data from AtoMx SIP (v1.3+) (@sec-export-raw-data)
2. Launch `napari-cosmx` within napari (@sec-launch-napari-cosmx)
3. Use the stitch widget to create napari-ready slide from raw data (@sec-stitch-images)

::: {.callout-note}
One of the main advantages of AtoMx SIP is that the data are stored for you. 
However, `napari-cosmx` currently requires the raw data downloaded and stored **locally**. Raw data can 
be quite large (100s of GBs per slide). It is possible
to store the data on a networked drive but we have noticed that stitching performance
is slower, depending on your network speed. Storing the data on a high-capacity and
fast I/O external hard drive may also be an option. 
:::



### 1. Export Raw Data {#sec-export-raw-data}

In AtoMx SIP, in the `Study details` panel on the upper right, click `Export` (@fig-atomx-export1). 

```{r}
#| eval: true
#| echo: false
#| label: "fig-atomx-export1"
#| fig-cap: "Click `Export` (available in AtoMx SIP version 1.3+)."

knitr::include_graphics("./figures/export-fig1.png")
```

Configure your export with the options indicated in @fig-atomx-export2 and click `EXPORT`. If you
would like to view metadata (optional). You can go ahead and download the Seurat data
now or as a separate step (_i.e._, Seurat data is not needed for stitching napari
files). 

```{r}
#| eval: true
#| echo: false
#| label: "fig-atomx-export2"
#| fig-cap: "Example configuration for exporting."

knitr::include_graphics("./figures/export-fig2.png")
```

When the export is ready, download the data. You can do this over the sftp 
protocol in a variety of application. Here, I'm using [Cyberduck](https://cyberduck.io/){target="_blank"} but you can use other programs. 

In Cyberduck, click `Open Connection`. In the dropdown menu, select `SFTP` and 
enter the URL, username, and your (AtoMx) password. Then click `Connect`. Example:
@fig-atomx-export3.

```{r}
#| eval: true
#| echo: false
#| label: "fig-atomx-export3"
#| fig-cap: "Example configuration for Cyberduck SFTP"

knitr::include_graphics("./figures/export-fig3.png")
```

Once connected, find the relevant folder, right click, and select `Download As...` (@fig-atomx-export4). 
Choose the location to on your computer to store the data. You may be able to store
it on a networked drive but this is currently untested. 

```{r}
#| eval: true
#| echo: false
#| label: "fig-atomx-export4"
#| fig-cap: "Download raw data somewhere on your desktop."

knitr::include_graphics("./figures/export-fig4.png")
```

Once downloaded and saved, follow the next step @sec-launch-napari-cosmx. 

### 2. Launch `napari-cosmx` {#sec-launch-napari-cosmx}

In order to use the stitching widget in the plugin, we must first launch the 
plugin. Currently, the only way to do that is to load an existing napari-ready 
folder. This can be any napari slide (_e.g._, a previous study or the two-FOV 
example described in @sec-napari-ready-files).

### 3. Stitching images {#sec-stitch-images}

Once `napari-cosmx` is launched, the stitching widget is located on the right-hand
panel (@fig-stitch-widget1). 

```{r}
#| eval: true
#| echo: false
#| label: "fig-stitch-widget1"
#| fig-cap: "Example showing a launched `napari-cosmx` plugin with a possibly-unrelated slide. To stitch, a new slide, Click `Browse...` in the `Stitch Images` widget located on the right-hand side of napari."

knitr::include_graphics("./figures/stitch-widget1.png")
```

We need to tell `napari-cosmx` where the raw data is located. In the right-hand panel
there is a `Stitch Images` widget. Click `Browse...` and navigate to the parent folder
containing the slide's raw data. In our minimum example above with two FOVs, I have named
the parent folder `two_fovs`. Your slide name will be labeled something else (@fig-stitch-widget2). 
Once you select `Open`, you should see the path to the raw data folder. If an unexpected format
was detected, there will be an error message (_e.g._, @fig-stitch-widget3).

```{r}
#| eval: true
#| echo: false
#| label: "fig-stitch-widget2"
#| fig-cap: "Example showing a launched `napari-cosmx` plugin with a possibly-unrelated slide. To stitch, a new slide, Click `Browse...` in the `Stitch Images` widget located on the right-hand side of napari."

knitr::include_graphics("./figures/stitch-widget2.png")
```

```{r}
#| eval: true
#| echo: false
#| label: "fig-stitch-widget3"
#| fig-cap: "Stitching widget prints the path to the correctly formatted raw data (left) or provides an error message if not formatted correctly (right). Note that only the correctly formatted location can proceed to the next step (choosing otuput folder)."

knitr::include_graphics("./figures/stitch-widget3.png")
```

Next, select the location where you want the plugin to return the napari-ready files.
It is recommend not to store it in the same location as the raw data. Here, I'm 
pointing it to a location adjacent to the raw data that I named `two_fovs_napari_files`
(@fig-stitch-widget4).

```{r}
#| eval: true
#| echo: false
#| label: "fig-stitch-widget4"
#| fig-cap: "Select output folder."

knitr::include_graphics("./figures/stitch-widget4.png")
```

Finally, click `Stitch`. Note: currently there is no refreshing or printing of
messages. Please do not click `Stitch` more than once. You may see napari become
unresponse, see the "spinning beachball" (Mac), _etc._ Once complete, you should
see messages that resemble that of @fig-stitch-widget5. If you see the last line
`See output folder for results`, you successfully converted the raw data into napari-ready files!

To view the results, simply close napari, reopen it, and drag your newly created results
into the application. 

```{r}
#| eval: true
#| echo: false
#| label: "fig-stitch-widget5"
#| fig-cap: "Select output folder."

knitr::include_graphics("./figures/stitch-widget5.png")
```



## Adding and viewing metadata {#sec-adding-metadata} 





