---
title: "FOV QC from single-cell gene expression in spatial dataset"
author:
  - name: Mithra Korukonda
    affiliations:
      - ref: nstg
      - ref: skorukonda
toc: true
toc-title: Contents
toc-depth: 3
toc-expand: 2
toc-location: left
date: "2024-05-20"
date-modified: "2024-05-20"
categories: [primary processing, quality control, registration]
draft: false
#image: figures/fov_qc_thumbnail.png
description: "Draft: Primary Registraton Quality Check" 
---

```{css}
#| echo: false
#| eval: true
.center {
  display: block;
  margin-left: auto;
  margin-right: auto;
  width: 50%;
}

```


# FOV QC

## Background

In most experiments, all FOVs will perform comparably, and data analyses need not consider FOV as a relevant variable.
However, FOVs can suffer a variety of technical effects, sometimes causing obvious bias in the 
data (e.g. all the cells in an FOV will be clustered as the same cell type), and sometimes more subtle. 
We recommend that FOV QC be performed early in analyses. Should misbehaving FOVs be detected,
we almost always recommend they be excluded. 

Here we'll describe known FOV-level artifacts, and we'll show use of R code for detecting 
impacted FOVs.

