<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.5.57">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Dan McGuire">
<meta name="dcterms.date" content="2024-06-19">
<meta name="description" content="Applications for visualization and cell typing using ‘smoothed’ marker genes based on expression nearest neighbors">

<title>Applications for visualization and cell typing using ‘smoothed’ marker genes – Blog</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../../site_libs/clipboard/clipboard.min.js"></script>
<script src="../../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../../site_libs/quarto-search/fuse.min.js"></script>
<script src="../../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../../">
<link href="../../assets/logo_Bruker_black.png" rel="icon" type="image/png">
<script src="../../site_libs/cookie-consent/cookie-consent.js"></script>
<link href="../../site_libs/cookie-consent/cookie-consent.css" rel="stylesheet">
<script src="../../site_libs/quarto-html/quarto.js"></script>
<script src="../../site_libs/quarto-html/popper.min.js"></script>
<script src="../../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../../site_libs/quarto-html/anchor.min.js"></script>
<link href="../../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="../../site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="../../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="../../site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>
<script async="" src="https://www.googletagmanager.com/gtag/js?id=G-29W4MW0Y2W"></script>

<script type="text/plain" cookie-consent="tracking">

window.dataLayer = window.dataLayer || [];
function gtag(){dataLayer.push(arguments);}
gtag('js', new Date());
gtag('config', 'G-29W4MW0Y2W', { 'anonymize_ip': true});
</script>

<script type="text/javascript" charset="UTF-8">
document.addEventListener('DOMContentLoaded', function () {
cookieconsent.run({
  "notice_banner_type":"simple",
  "consent_type":"implied",
  "palette":"light",
  "language":"en",
  "page_load_consent_levels":["strictly-necessary","functionality","tracking","targeting"],
  "notice_banner_reject_button_hide":false,
  "preferences_center_close_button_hide":false,
  "website_name":""
  ,
"language":"en"
  });
});
</script> 
  

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top quarto-banner">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a href="../../index.html" class="navbar-brand navbar-brand-logo">
    <img src="../.././assets/logo_Bruker_white.png" alt="" class="navbar-logo">
    </a>
    <a class="navbar-brand" href="../../index.html">
    <span class="navbar-title">Blog</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" role="menu" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../license.html"> 
<span class="menu-text">License</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="../../link-to-code.html"> 
<span class="menu-text">Code</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="../../about.html"> 
<span class="menu-text">Browse Posts by Topic</span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://github.com/Nanostring-Biostats/CosMx-Analysis-Scratch-Space"> <i class="bi bi-github" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://twitter.com/nanostringtech"> <i class="bi bi-twitter" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="https://www.linkedin.com/company/bruker-spatial"> <i class="bi bi-linkedin" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
  <li class="nav-item compact">
    <a class="nav-link" href="../../index.xml"> <i class="bi bi-rss" role="img">
</i> 
<span class="menu-text"></span></a>
  </li>  
</ul>
          </div> <!-- /navcollapse -->
            <div class="quarto-navbar-tools">
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
      </div> <!-- /container-fluid -->
    </nav>
</header>
<!-- content -->
<header id="title-block-header" class="quarto-title-block default toc-left page-columns page-full">
  <div class="quarto-title-banner page-columns page-full">
    <div class="quarto-title column-body">
      <div class="quarto-title-block"><div><h1 class="title">Applications for visualization and cell typing using ‘smoothed’ marker genes</h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
                  <div>
        <div class="description">
          Applications for visualization and cell typing using ‘smoothed’ marker genes based on expression nearest neighbors
        </div>
      </div>
                          <div class="quarto-categories">
                <div class="quarto-category">visualization</div>
                <div class="quarto-category">cell typing</div>
              </div>
                  </div>
  </div>
    
  <div class="quarto-title-meta-author">
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-heading">Affiliations</div>
    
      <div class="quarto-title-meta-contents">
      <p class="author">Dan McGuire <a href="https://orcid.org/0009-0006-0286-9625" class="quarto-title-author-orcid"> <img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAYAAAAf8/9hAAAAGXRFWHRTb2Z0d2FyZQBBZG9iZSBJbWFnZVJlYWR5ccllPAAAA2ZpVFh0WE1MOmNvbS5hZG9iZS54bXAAAAAAADw/eHBhY2tldCBiZWdpbj0i77u/IiBpZD0iVzVNME1wQ2VoaUh6cmVTek5UY3prYzlkIj8+IDx4OnhtcG1ldGEgeG1sbnM6eD0iYWRvYmU6bnM6bWV0YS8iIHg6eG1wdGs9IkFkb2JlIFhNUCBDb3JlIDUuMC1jMDYwIDYxLjEzNDc3NywgMjAxMC8wMi8xMi0xNzozMjowMCAgICAgICAgIj4gPHJkZjpSREYgeG1sbnM6cmRmPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5LzAyLzIyLXJkZi1zeW50YXgtbnMjIj4gPHJkZjpEZXNjcmlwdGlvbiByZGY6YWJvdXQ9IiIgeG1sbnM6eG1wTU09Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC9tbS8iIHhtbG5zOnN0UmVmPSJodHRwOi8vbnMuYWRvYmUuY29tL3hhcC8xLjAvc1R5cGUvUmVzb3VyY2VSZWYjIiB4bWxuczp4bXA9Imh0dHA6Ly9ucy5hZG9iZS5jb20veGFwLzEuMC8iIHhtcE1NOk9yaWdpbmFsRG9jdW1lbnRJRD0ieG1wLmRpZDo1N0NEMjA4MDI1MjA2ODExOTk0QzkzNTEzRjZEQTg1NyIgeG1wTU06RG9jdW1lbnRJRD0ieG1wLmRpZDozM0NDOEJGNEZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wTU06SW5zdGFuY2VJRD0ieG1wLmlpZDozM0NDOEJGM0ZGNTcxMUUxODdBOEVCODg2RjdCQ0QwOSIgeG1wOkNyZWF0b3JUb29sPSJBZG9iZSBQaG90b3Nob3AgQ1M1IE1hY2ludG9zaCI+IDx4bXBNTTpEZXJpdmVkRnJvbSBzdFJlZjppbnN0YW5jZUlEPSJ4bXAuaWlkOkZDN0YxMTc0MDcyMDY4MTE5NUZFRDc5MUM2MUUwNEREIiBzdFJlZjpkb2N1bWVudElEPSJ4bXAuZGlkOjU3Q0QyMDgwMjUyMDY4MTE5OTRDOTM1MTNGNkRBODU3Ii8+IDwvcmRmOkRlc2NyaXB0aW9uPiA8L3JkZjpSREY+IDwveDp4bXBtZXRhPiA8P3hwYWNrZXQgZW5kPSJyIj8+84NovQAAAR1JREFUeNpiZEADy85ZJgCpeCB2QJM6AMQLo4yOL0AWZETSqACk1gOxAQN+cAGIA4EGPQBxmJA0nwdpjjQ8xqArmczw5tMHXAaALDgP1QMxAGqzAAPxQACqh4ER6uf5MBlkm0X4EGayMfMw/Pr7Bd2gRBZogMFBrv01hisv5jLsv9nLAPIOMnjy8RDDyYctyAbFM2EJbRQw+aAWw/LzVgx7b+cwCHKqMhjJFCBLOzAR6+lXX84xnHjYyqAo5IUizkRCwIENQQckGSDGY4TVgAPEaraQr2a4/24bSuoExcJCfAEJihXkWDj3ZAKy9EJGaEo8T0QSxkjSwORsCAuDQCD+QILmD1A9kECEZgxDaEZhICIzGcIyEyOl2RkgwAAhkmC+eAm0TAAAAABJRU5ErkJggg=="></a></p>
    </div>
    <div class="quarto-title-meta-contents">
          <p class="affiliation">
              Bruker Spatial Biology
            </p>
          <p class="affiliation">
              Github: <a href="https://github.com/dan11mcguire" target="_blank">dan11mcguire</a>
            </p>
        </div>
    </div>

  <div class="quarto-title-meta">

        
      <div>
      <div class="quarto-title-meta-heading">Published</div>
      <div class="quarto-title-meta-contents">
        <p class="date">June 19, 2024</p>
      </div>
    </div>
    
      
    </div>
    
  
  </header><div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <nav id="TOC" role="doc-toc" class="toc-active" data-toc-expanded="2">
    <h2 id="toc-title">Contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction"><span class="header-section-number">1</span> Introduction</a></li>
  <li><a href="#a-motivating-example-noisy-t-cell-typing-in-non-small-cell-lung-cancer-tissue" id="toc-a-motivating-example-noisy-t-cell-typing-in-non-small-cell-lung-cancer-tissue" class="nav-link" data-scroll-target="#a-motivating-example-noisy-t-cell-typing-in-non-small-cell-lung-cancer-tissue"><span class="header-section-number">2</span> A motivating example: Noisy T-cell typing in non-small cell lung cancer tissue</a></li>
  <li><a href="#using-smoothed-foxp3-to-highlight-treg-focal-point-and-refine-cell-types." id="toc-using-smoothed-foxp3-to-highlight-treg-focal-point-and-refine-cell-types." class="nav-link" data-scroll-target="#using-smoothed-foxp3-to-highlight-treg-focal-point-and-refine-cell-types."><span class="header-section-number">3</span> Using ‘smoothed’ FOXP3 to highlight Treg focal point and refine cell types.</a></li>
  <li><a href="#other-marker-genes" id="toc-other-marker-genes" class="nav-link" data-scroll-target="#other-marker-genes"><span class="header-section-number">4</span> Other marker genes</a></li>
  <li><a href="#conclusion" id="toc-conclusion" class="nav-link" data-scroll-target="#conclusion"><span class="header-section-number">5</span> Conclusion</a></li>
  </ul>
</nav>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
    </div>
<!-- main -->
<main class="content quarto-banner-title-block" id="quarto-document-content">





<section id="introduction" class="level1" data-number="1">
<h1 data-number="1"><span class="header-section-number">1</span> Introduction</h1>
<p>Marker genes are genes that are expressed primarily within a single cell type, and are often used to delineate and label clusters during cell typing in a scRNAseq analysis.</p>
<p>With spatially-resolved transcriptomics (SRT) data, a number of factors can contribute to challenges in a cell typing analysis, as well as hinder our ability to visualize our favorite marker genes. For example, these factors may include lower sensitivity compared to scRNAseq data, background due to autoflourescence, and segmentation error.</p>
<p>A common feedback for new SRT analysts when dealing with SRT data may be along the lines of “Why don’t I see gene ‘X’ in a majority of cells for cell type ‘Y’?”.</p>
<p>In this post we’ll discuss:</p>
<ul>
<li>Why counts of a single marker gene are not definitive of cell type</li>
<li>How to derive more useful / less noisy / “smoothed” expression of marker genes</li>
<li>How to perform fine-grained subtyping using smoothed marker genes, with T-cell subtyping as motivation</li>
</ul>
</section>
<section id="a-motivating-example-noisy-t-cell-typing-in-non-small-cell-lung-cancer-tissue" class="level1" data-number="2">
<h1 data-number="2"><span class="header-section-number">2</span> A motivating example: Noisy T-cell typing in non-small cell lung cancer tissue</h1>
<p>For example, here is a data set below consisting of non-small cell lung cancer tissues. Let’s load it in, run some quick unrefined-cell typing using <a href="https://github.com/Nanostring-Biostats/InSituType?tab=readme-ov-file" target="_blank">InSituType</a> <span class="citation" data-cites="Danaher2022">(<a href="#ref-Danaher2022" role="doc-biblioref">Danaher et al. 2022</a>)</span>, and take a look at the initial results.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(InSituType)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"seurat_object.Rds"</span>)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="do">### semi-supervised cell typing with 3 unsupervised clusters, using 'ioprofiles' reference matrix</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>insitu <span class="ot">&lt;-</span> </span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>InSituType<span class="sc">::</span><span class="fu">insitutype</span>(Matrix<span class="sc">::</span><span class="fu">t</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts)</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">neg =</span> Matrix<span class="sc">::</span><span class="fu">colMeans</span>(sem[[<span class="st">"negprobes"</span>]])</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">reference_profiles =</span> InSituType<span class="sc">::</span>ioprofiles</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">n_clusts =</span> <span class="dv">3</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="do">### clean up cell type names and save results back to seurat object</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>clustdt <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">clust_o =</span> insitu<span class="sc">$</span>clust, <span class="at">row.names=</span><span class="fu">names</span>(insitu<span class="sc">$</span>clust))</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a>clustdt[,clust<span class="sc">:</span><span class="er">=</span>clust_o]</span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>clustdt[<span class="fu">grep</span>(<span class="st">"T CD4"</span>, clust),clust<span class="sc">:</span><span class="er">=</span><span class="st">"T CD4"</span>]</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>clustdt[<span class="fu">grep</span>(<span class="st">"T CD8"</span>, clust),clust<span class="sc">:</span><span class="er">=</span><span class="st">"T CD8"</span>]</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>clustdt[<span class="fu">grep</span>(<span class="st">"B-cell"</span>, clust),clust<span class="sc">:</span><span class="er">=</span><span class="st">"B cell"</span>]</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a>clustdt[,.N,by<span class="ot">=</span>.(clust)]</span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>sem<span class="sc">@</span>meta.data<span class="sc">$</span>clust <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> </span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>Seurat<span class="sc">::</span><span class="fu">AddMetaData</span>(sem</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>                    ,<span class="fu">data.frame</span>(clustdt[,.(clust, clust_o)]</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>                                ,<span class="at">row.names=</span><span class="fu">names</span>(insitu<span class="sc">$</span>clust)</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>                                )</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>                    )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here we can plot the InSituType clusters (‘clust’) on the UMAP. This UMAP was created using ‘Analytic Pearson residuals’ <span class="citation" data-cites="Lause2021">(<a href="#ref-Lause2021" role="doc-biblioref">Lause, Berens, and Kobak 2021</a>)</span>.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>umapf <span class="ot">&lt;-</span> <span class="cf">function</span>(umapreduc</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                  ,clustercol</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                  ,semuse</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                  ,<span class="at">cls=</span><span class="cn">NULL</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                  ,<span class="at">xlim =</span> <span class="cn">NULL</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>                  ,<span class="at">ylim =</span> <span class="cn">NULL</span>){</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  umapd <span class="ot">&lt;-</span>  </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>(semuse<span class="sc">@</span>reductions[[umapreduc]]<span class="sc">@</span>cell.embeddings</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">keep.rownames =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(umapd, <span class="fu">c</span>(<span class="fu">names</span>(umapd)[<span class="dv">2</span><span class="sc">:</span><span class="dv">3</span>]), <span class="fu">c</span>(<span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>))</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  obsmrk <span class="ot">&lt;-</span> <span class="fu">merge</span>(<span class="fu">data.table</span>(semuse<span class="sc">@</span>meta.data), umapd</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>                  ,<span class="at">by.x=</span><span class="st">"cell_ID"</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>                  , <span class="at">by.y=</span><span class="st">"rn"</span>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>  obstxt <span class="ot">&lt;-</span> obsmrk[,<span class="fu">lapply</span>(.SD, median),by<span class="ot">=</span><span class="fu">c</span>(clustercol),.SDcols<span class="ot">=</span><span class="fu">paste0</span>(<span class="st">"UMAP_"</span>,<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>)]</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> </span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggplot</span>(obsmrk, <span class="fu">aes</span>(UMAP_1, UMAP_2, <span class="at">color=</span>.data[[clustercol]])) <span class="sc">+</span> </span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>      <span class="fu">theme_bw</span>() <span class="sc">+</span> <span class="fu">coord_fixed</span>(<span class="at">xlim=</span>xlim, <span class="at">ylim=</span>ylim) <span class="sc">+</span> </span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>      <span class="fu">geom_label_repel</span>(<span class="at">data=</span>obstxt, <span class="fu">aes</span>(<span class="at">x=</span>UMAP_1, <span class="at">y=</span>UMAP_2, <span class="at">label=</span>.data[[clustercol]]),<span class="at">show.legend=</span><span class="cn">FALSE</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>                       ,<span class="at">inherit.aes=</span><span class="cn">FALSE</span>,<span class="at">color=</span><span class="st">'black'</span>)</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">is.null</span>(cls)){</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> p <span class="sc">+</span>  </span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_manual</span>(<span class="at">values=</span><span class="fu">rep</span>(<span class="fu">unname</span>(pals<span class="sc">::</span><span class="fu">alphabet</span>()), <span class="dv">3</span>)</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>                         ,<span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>)))</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> p <span class="sc">+</span>  </span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>      <span class="fu">scale_color_manual</span>(<span class="at">values=</span>cls</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>                         ,<span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>)))</span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ctpal <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'#C20088'</span>,<span class="st">'#005C31'</span>,<span class="st">'#2BCE48'</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>           ,<span class="st">'#4C005C'</span>,<span class="st">'#F0A0FF'</span>,<span class="st">'#003380'</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>           ,<span class="st">'#FFCC99'</span>,<span class="st">'#8F7C00'</span>,<span class="st">'#9DCC00'</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>           ,<span class="st">'#191919'</span>,<span class="st">'#94FFB5'</span>,<span class="st">'#0075DC'</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>           ,<span class="st">'#FFA8BB'</span>,<span class="st">'#FFA405'</span>,<span class="st">'#993F00'</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>           ,<span class="st">'#808080'</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(ctpal) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">'endothelial'</span>,<span class="st">'mDC'</span>,<span class="st">'plasmablast'</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>                  ,<span class="st">'b'</span>,<span class="st">'B cell'</span>,<span class="st">'pDC'</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>                  ,<span class="st">'macrophage'</span>,<span class="st">'a'</span>,<span class="st">'mast'</span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a>                  ,<span class="st">'T CD4'</span>,<span class="st">'fibroblast'</span>,<span class="st">'T CD8'</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>                  ,<span class="st">'NK'</span>,<span class="st">'Treg'</span>,<span class="st">'neutrophil'</span></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>                  ,<span class="st">'c'</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a><span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"clust"</span>, sem, <span class="at">cls =</span> ctpal)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/umap_clust_all.png" class="img-fluid figure-img" width="2661"></p>
</figure>
</div>
</div>
</div>
<p>Let’s zoom in on some of the supervised clusters we hoped to identify below. We can see that some of the major cell types are somewhat clearly delineated on the UMAP.</p>
<ul>
<li>Lymphocytes; T-cell types (T CD8, T CD4, and Treg) are clustered together, and near the ‘B cell’ cluster.</li>
<li>Myeloid cell types; (macrophage, pDC, mDC) are clustered in a similar area.</li>
</ul>
<p>But there does appear to potentially be some noise in delineating cell types within those major categories.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"clust"</span>, sem, <span class="at">cls =</span> ctpal, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/umap_clust_imm.png" class="img-fluid figure-img" width="2214"></p>
</figure>
</div>
</div>
</div>
<p>We can focus on T cells in this dataset as a driving example for using smoothing as an approach for addressing challenging cell typing and visualization problems.</p>
<p>Canonical marker genes for T cells include</p>
<ul>
<li><em>CD3</em> (expected to be expressed in all T-cell types),</li>
<li><em>CD4</em> (commonly used together with <em>CD3</em> to identify T CD4 cells. <em>CD4</em> can also be expressed in myeloid cells.)</li>
<li><em>FOXP3</em> (commonly used together with <em>CD3</em> to identify Treg cells. Treg cells are a special subset of T CD4 cells)</li>
<li><em>CD8A</em> and <em>CD8B</em> (used together with <em>CD3</em> to identify T CD8 cells)</li>
</ul>
<p>First, let’s take a look at the relative frequency of our T-cell clusters ‘Treg’, ‘T CD8’, and ‘T CD4’. Tregs are supposed to be a rare sub type of T CD4 cells. Yet we have more than 3x as many Tregs called than we do T CD4.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>bard <span class="ot">&lt;-</span> <span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)[,.N,by<span class="ot">=</span>.(clust)]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>bard[,clust<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(clust, <span class="at">levels=</span>bard[<span class="fu">order</span>(<span class="sc">-</span>N),clust])]</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bard[<span class="fu">grep</span>(<span class="st">"^T"</span>,clust)], <span class="fu">aes</span>(<span class="at">x=</span>clust, <span class="at">y=</span>N,<span class="at">fill=</span>clust)) <span class="sc">+</span> </span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">16</span>)) <span class="sc">+</span> </span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span>ctpal, <span class="at">guide =</span> <span class="fu">guide_legend</span>(<span class="at">reverse=</span><span class="cn">TRUE</span>)) <span class="sc">+</span> </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">n.breaks=</span><span class="dv">12</span>, <span class="at">labels =</span> scales<span class="sc">::</span>comma, <span class="at">name =</span> <span class="st">"# of cells"</span>) <span class="sc">+</span> </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"# of cells called by T cell subtype"</span>) <span class="sc">+</span> </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>clust, <span class="at">y=</span>N <span class="sc">+</span> <span class="dv">200</span>, <span class="at">label=</span>scales<span class="sc">::</span><span class="fu">comma</span>(N))) <span class="sc">+</span> </span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/tcell_calls_barplot.png" class="img-fluid figure-img" width="2298"></p>
</figure>
</div>
</div>
</div>
<p>Do we really have this many Treg cells in our data?</p>
<p>To gather more evidence toward answering this kind of question, we can calculate the proportion of cells in each cell type expressing a particular marker gene, along with a fold change comparison for each cluster relative to other clusters.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>totalcount_norm <span class="ot">&lt;-</span> <span class="cf">function</span>(sm){</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>    libsizes <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">colSums</span>(sm)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>    scalefactor <span class="ot">&lt;-</span> <span class="fu">mean</span>(libsizes)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>    libsizes[libsizes<span class="sc">==</span><span class="dv">0</span>] <span class="ot">&lt;-</span> <span class="dv">1</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>    normed <span class="ot">&lt;-</span> sm <span class="sc">%*%</span> Matrix<span class="sc">::</span><span class="fu">Diagonal</span>(<span class="at">x=</span>scalefactor<span class="sc">/</span>libsizes)</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">dimnames</span>(normed) <span class="ot">&lt;-</span> <span class="fu">dimnames</span>(sm)</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">return</span>(normed)</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a>clusterwise_fold_change_stats <span class="ot">&lt;-</span> <span class="cf">function</span>(<span class="at">cnts=</span><span class="cn">NULL</span>, <span class="at">normed =</span> <span class="cn">NULL</span>, metainfo, clustercol){</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(<span class="fu">missing</span>(normed)){</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>    normed <span class="ot">&lt;-</span> <span class="fu">totalcount_norm</span>(cnts) </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  outl <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(ii <span class="cf">in</span> <span class="fu">unique</span>(metainfo[[clustercol]])){</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>    cells_ii <span class="ot">&lt;-</span> metainfo[metainfo[[clustercol]]<span class="sc">==</span>ii,cell_ID]</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>    cells_iiprime <span class="ot">&lt;-</span> metainfo[metainfo[[clustercol]]<span class="sc">!=</span>ii,cell_ID]</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>    cluster_expr_ii  <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">rowMeans</span>(normed[,cells_ii,<span class="at">drop=</span><span class="cn">FALSE</span>])</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>    cluster_expr_iiprime <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">rowMeans</span>(normed[,cells_iiprime,<span class="at">drop=</span><span class="cn">FALSE</span>]) </span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>    cluster_prop_ii <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">rowMeans</span>(normed[,cells_ii,<span class="at">drop=</span><span class="cn">FALSE</span>] <span class="sc">&gt;</span> <span class="dv">0</span>) </span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>    cluster_prop_iiprime <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">rowMeans</span>(normed[,cells_iiprime,<span class="at">drop=</span><span class="cn">FALSE</span>] <span class="sc">&gt;</span> <span class="dv">0</span>) </span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>    fctbl <span class="ot">&lt;-</span> <span class="fu">data.table</span>(<span class="at">cluster=</span>ii</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true" tabindex="-1"></a>                        ,<span class="at">cluster_expr =</span> cluster_expr_ii</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true" tabindex="-1"></a>                        ,<span class="at">clusterprime_expr =</span> cluster_expr_iiprime</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true" tabindex="-1"></a>                        ,<span class="at">gene =</span> <span class="fu">names</span>(cluster_expr_ii)</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true" tabindex="-1"></a>                        ,<span class="at">cluster_prop =</span> cluster_prop_ii</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true" tabindex="-1"></a>                        ,<span class="at">clusterprime_prop =</span> cluster_prop_iiprime</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true" tabindex="-1"></a>    )[,typ<span class="sc">:</span><span class="er">=</span>clustercol]</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true" tabindex="-1"></a>    fctbl[,fold_change<span class="sc">:</span><span class="er">=</span>cluster_expr <span class="sc">/</span> clusterprime_expr]</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true" tabindex="-1"></a>    fctbl[,fold_change_prop<span class="sc">:</span><span class="er">=</span>cluster_prop <span class="sc">/</span> clusterprime_prop]</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true" tabindex="-1"></a>    outl[[<span class="fu">paste0</span>(ii)]] <span class="ot">&lt;-</span> <span class="fu">copy</span>(fctbl) </span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true" tabindex="-1"></a>  } </span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(<span class="at">fc =</span> <span class="fu">rbindlist</span>(outl))</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We can see that <em>FOXP3</em> is expressed in a much larger frequency in Treg cells relative to other T-cell subtypes– this is good.</p>
<p>However, only 12.7% of our Treg cells are <em>FOXP3</em> positive, which may be a concern given that we have so many more Treg cells than other T-cell subtypes.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>cluster_level_stats <span class="ot">&lt;-</span>  </span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a><span class="fu">clusterwise_fold_change_stats</span>(<span class="at">cnts =</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                   ,<span class="at">metainfo =</span> <span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                   ,<span class="at">clustercol =</span> <span class="st">"clust"</span>)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>bard <span class="ot">&lt;-</span> cluster_level_stats[gene<span class="sc">==</span><span class="st">"FOXP3"</span>][<span class="fu">grep</span>(<span class="st">"^T"</span>,cluster)]</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>bard[,cluster<span class="sc">:</span><span class="er">=</span><span class="fu">factor</span>(cluster, <span class="at">levels=</span>bard[<span class="fu">order</span>(<span class="sc">-</span>cluster_prop),cluster])]</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(bard, <span class="fu">aes</span>(<span class="at">x=</span>cluster, <span class="at">y=</span>cluster_prop, <span class="at">fill=</span>cluster)) <span class="sc">+</span> </span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">text=</span><span class="fu">element_text</span>(<span class="at">size=</span><span class="dv">16</span>)) <span class="sc">+</span> </span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>) <span class="sc">+</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values=</span>ctpal, <span class="at">guide =</span> <span class="fu">guide_legend</span>(<span class="at">reverse=</span><span class="cn">TRUE</span>)) <span class="sc">+</span> </span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">n.breaks=</span><span class="dv">12</span>, <span class="at">labels =</span> scales<span class="sc">::</span>comma, <span class="at">name =</span> <span class="st">"Proportion of cells expressing the FOXP3 marker"</span>) <span class="sc">+</span> </span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"FOXP3"</span>) <span class="sc">+</span> </span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>cluster, <span class="at">y=</span>cluster_prop <span class="sc">+</span> <span class="fl">0.01</span>, <span class="at">label=</span><span class="fu">formatC</span>(cluster_prop,<span class="dv">3</span>))) <span class="sc">+</span> </span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_flip</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/tcell_foxp3_pct_barplot.png" class="img-fluid figure-img" width="2298"></p>
</figure>
</div>
</div>
</div>
<p>At this stage, one may reasonably wonder: “If we define a Treg cell based on <em>FOXP3</em> expression, why not call a cell”Treg” <strong>only</strong> if it is expressing <em>FOXP3</em>?”</p>
<p>Here we can illustrate the challenge this would present by plotting <em>FOXP3</em> expression on the UMAP. On the one hand, if we look closely, we can see what looks like might be a “hot spot” on the UMAP where we are calling Treg cells, and we also have higher <em>FOXP3</em> expression.</p>
<p>But setting a threshold based on <em>FOXP3</em> expression to define our Treg cells would be futile.</p>
<p>We have many cells near our ‘<em>FOXP3</em> hotspot’ which may very well be Tregs, but do not have any counts of the marker gene. On the other hand, we can see cells of all types, all over the UMAP, which have one or several counts of <em>FOXP3</em> expressed possibly due to some form of background.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>marker_umap_plot <span class="ot">&lt;-</span> <span class="cf">function</span>(marker_matrix, marker, sem</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>                             ,umapreduc, <span class="at">scale_expr =</span> <span class="cn">TRUE</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>                             ,<span class="at">xlim=</span><span class="cn">NULL</span>,<span class="at">ylim=</span><span class="cn">NULL</span>){</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  pd <span class="ot">&lt;-</span>  </span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>(sem<span class="sc">@</span>reductions[[umapreduc]]<span class="sc">@</span>cell.embeddings</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">keep.rownames =</span> <span class="cn">TRUE</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(pd, <span class="fu">c</span>(<span class="fu">names</span>(pd)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]), <span class="fu">c</span>(<span class="st">"cell_ID"</span>, <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>))</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>  pd[<span class="fu">match</span>(<span class="fu">colnames</span>(marker_matrix), cell_ID),(<span class="fu">c</span>(marker))<span class="sc">:</span><span class="er">=</span>marker_matrix[marker,]]</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setkeyv</span>(pd, marker)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(scale_expr){</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> </span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggplot</span>(pd </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>             ,<span class="fu">aes</span>(<span class="at">x=</span>UMAP_1, <span class="at">y =</span> UMAP_2, <span class="at">color=</span><span class="fu">scale</span>(.data[[marker]]))) </span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> </span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggplot</span>(pd </span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>             ,<span class="fu">aes</span>(<span class="at">x=</span>UMAP_1, <span class="at">y =</span> UMAP_2, <span class="at">color=</span>.data[[marker]])) </span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradientn</span>(<span class="at">colors=</span><span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Reds"</span>)) <span class="sc">+</span> </span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">n.breaks=</span><span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">n.breaks=</span><span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span>marker) <span class="sc">+</span> <span class="fu">coord_fixed</span>(<span class="at">xlim=</span>xlim,<span class="at">ylim=</span>ylim) </span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>raw_foxp3_plot <span class="ot">&lt;-</span> </span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">marker_umap_plot</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts, <span class="st">"FOXP3"</span>, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>                   ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>),<span class="at">scale=</span><span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"raw FOXP3"</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>ctpal_sub <span class="ot">&lt;-</span> ctpal</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>ctpal_sub[<span class="fu">grep</span>(<span class="st">"^T"</span>,<span class="fu">names</span>(ctpal_sub),<span class="at">invert=</span><span class="cn">TRUE</span>)] <span class="ot">&lt;-</span> <span class="st">"grey77"</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>  raw_foxp3_plot</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a>  ,<span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"clust"</span>, sem</span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">cls =</span> ctpal_sub, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>))</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">nrow=</span><span class="dv">2</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/foxp3raw.png" class="img-fluid figure-img" width="1641"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="using-smoothed-foxp3-to-highlight-treg-focal-point-and-refine-cell-types." class="level1" data-number="3">
<h1 data-number="3"><span class="header-section-number">3</span> Using ‘smoothed’ FOXP3 to highlight Treg focal point and refine cell types.</h1>
<p>One potential approach to cleaning this up is to use some form of ‘smoothed’ marker gene expression, rather than the raw <em>FOXP3</em> counts.</p>
<p>We can describe the calculation for ‘smoothed’ marker expression in two basic steps:</p>
<ol type="1">
<li>Find the <span class="math inline">\(K\)</span> nearest neighbors of each cell in UMAP space.</li>
<li>Average the raw (or normalized) expression of the gene across that cells nearest neighbors.</li>
</ol>
<p>There’s a few different ways we might think about motivating this approach. The UMAP is an approximate manifold projection, which takes a low (2-d) representation of our high-dimensional (960 genes) data set. The manifold is ‘locally connected’, meaning cells near each other in UMAP space are also similar to each other in expression space. By averaging the expression of a marker gene of interest across cells with similar profiles, we can make a simple imputation of what we’d expect to see for the marker gene, given other cells of similar expression profiles.</p>
<p>This function below identifies the nearest neighbors in <em>UMAP</em> space, and makes a smoothing matrix, which can be used to get the average expression of any gene among a given cell’s nearest neighbors. Here is a simple implementation of the function, to take a ‘Seurat’ object as input, but could easily be modified for other data formats.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="do">## create a column-standardized nearest neighbor matrix </span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="do">## based on nearest neighbors in umap coordinates</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param sem a seureat object, which contains the UMAP reduction.</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param umapreduc name of the umap reduction, i.e. "umap"</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#' @param n_neighbor number of nearest neighbors to be identified.</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">#' @return "smoother", a smoothing matrix. expression (genes x cells ) x  smoother(cells x cells) , </span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a><span class="co">#' will create an averaged expression across nearest neighbors.</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a><span class="co">#' </span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>umap_nn <span class="ot">&lt;-</span> <span class="cf">function</span>(sem, umapreduc, <span class="at">n_neighbors=</span><span class="dv">100</span>){</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>  <span class="do">## extract umap coordinates </span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>  umapd <span class="ot">&lt;-</span>  </span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>(sem<span class="sc">@</span>reductions[[umapreduc]]<span class="sc">@</span>cell.embeddings</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">keep.rownames =</span> <span class="cn">TRUE</span>)</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(umapd, <span class="fu">c</span>(<span class="fu">names</span>(umapd)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]), <span class="fu">c</span>(<span class="st">"cell_ID"</span>, <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>))</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>  <span class="do">## identify nearest n_neighbors (+1 includes the cell as a neighbor to itself)</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>  nn_umap <span class="ot">&lt;-</span> RANN<span class="sc">::</span><span class="fu">nn2</span>(umapd[,.(UMAP_1, UMAP_2)],<span class="at">k =</span> n_neighbors <span class="sc">+</span> <span class="dv">1</span>)<span class="sc">$</span>nn.idx</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>  nn_umap <span class="ot">&lt;-</span> data.table<span class="sc">::</span><span class="fu">melt</span>(<span class="fu">cbind</span>(umapd[,.(cell_ID)], <span class="fu">data.table</span>(nn_umap))</span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a>                              , <span class="at">id.vars=</span><span class="fu">c</span>(<span class="st">"cell_ID"</span>, <span class="st">"V1"</span>))</span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">colnames</span>(nn_umap) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"cell_ID1"</span>, <span class="st">"cell_ID1_idx"</span>, <span class="st">"neighbor"</span>, <span class="st">"cell_ID2_idx"</span>)</span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>  nn_umap <span class="ot">&lt;-</span> <span class="fu">merge</span>(nn_umap</span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>                   , nn_umap[,.(<span class="at">cell_ID2=</span>cell_ID1, <span class="at">cell_ID2_idx=</span>cell_ID1_idx)][</span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>                     ,<span class="fu">unique</span>(.SD)],<span class="at">by=</span><span class="st">"cell_ID2_idx"</span>)</span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Cell x cell neighbor indicator matrix</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>  wumap <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">sparseMatrix</span>(<span class="at">i =</span> <span class="fu">c</span>(<span class="fu">unique</span>(nn_umap<span class="sc">$</span>cell_ID1_idx), nn_umap<span class="sc">$</span>cell_ID2_idx)</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>                                ,<span class="at">j=</span><span class="fu">c</span>(<span class="fu">unique</span>(nn_umap<span class="sc">$</span>cell_ID1_idx), nn_umap<span class="sc">$</span>cell_ID1_idx)</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>                                ,<span class="at">x=</span><span class="dv">1</span>)</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames</span>(wumap) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(wumap) <span class="ot">&lt;-</span> nn_umap[<span class="fu">order</span>(nn_umap<span class="sc">$</span>cell_ID1_idx),<span class="fu">unique</span>(cell_ID1)]</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>  <span class="do">## Column standardize, so that columns (cells) sum to 1, and each neighbor given equal weight.</span></span>
<span id="cb10-36"><a href="#cb10-36" aria-hidden="true" tabindex="-1"></a>  mumap <span class="ot">&lt;-</span> Matrix<span class="sc">::</span><span class="fu">sparseMatrix</span>(<span class="at">i=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(wumap)</span>
<span id="cb10-37"><a href="#cb10-37" aria-hidden="true" tabindex="-1"></a>                                ,<span class="at">j=</span><span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(wumap)</span>
<span id="cb10-38"><a href="#cb10-38" aria-hidden="true" tabindex="-1"></a>                                ,<span class="at">x=</span><span class="dv">1</span><span class="sc">/</span>Matrix<span class="sc">::</span><span class="fu">colSums</span>(wumap)</span>
<span id="cb10-39"><a href="#cb10-39" aria-hidden="true" tabindex="-1"></a>                                )</span>
<span id="cb10-40"><a href="#cb10-40" aria-hidden="true" tabindex="-1"></a>  <span class="fu">dimnames</span>(mumap) <span class="ot">&lt;-</span> <span class="fu">dimnames</span>(wumap)</span>
<span id="cb10-41"><a href="#cb10-41" aria-hidden="true" tabindex="-1"></a>  smoother <span class="ot">&lt;-</span> mumap <span class="sc">%*%</span> wumap</span>
<span id="cb10-42"><a href="#cb10-42" aria-hidden="true" tabindex="-1"></a>  smoother <span class="ot">&lt;-</span> smoother[,<span class="fu">colnames</span>(sem)]</span>
<span id="cb10-43"><a href="#cb10-43" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(smoother)</span>
<span id="cb10-44"><a href="#cb10-44" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Here, we can apply our ‘smoothing matrix’ to some T-cell marker genes and compare the difference on the UMAP. Whereas raw marker expression is noisy and sparse, smoothed expression clearly highlights the focal points for each marker on the UMAP.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>smoother <span class="ot">&lt;-</span> <span class="fu">umap_nn</span>(sem, <span class="st">"pearsonumap"</span>, <span class="at">n_neighbors =</span> <span class="dv">500</span>)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> <span class="fu">SetAssayData</span>(sem, <span class="st">"data"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">new.data =</span> <span class="fu">totalcount_norm</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>smoothed_tcell_markers <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data[<span class="fu">c</span>(<span class="st">"FOXP3"</span>,<span class="st">"CD3G"</span>,<span class="st">"CD8B"</span>, <span class="st">"CD4"</span>),] <span class="sc">%*%</span> smoother</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(g <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"FOXP3"</span>, <span class="st">"CD8B"</span>,<span class="st">"CD3G"</span>)){</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(g,<span class="st">".smooth"</span>)]] <span class="ot">&lt;-</span> </span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">marker_umap_plot</span>(smoothed_tcell_markers, g, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>)) <span class="sc">+</span> </span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"smooth "</span>,g))</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(g,<span class="st">".raw"</span>)]] <span class="ot">&lt;-</span> </span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">marker_umap_plot</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts, g, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>)) <span class="sc">+</span> </span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"raw "</span>,g))</span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist=</span>plist, <span class="at">nrow=</span><span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/smooth_vs_raw.png" class="img-fluid figure-img" width="2050"></p>
</figure>
</div>
</div>
</div>
<p>From here, we could take a number of different approaches to ‘clean up’ our T-cell calls. For example, because Tregs are a subset of T CD4 cells, we might consider reassigning them to T CD4 if they have low ‘smoothed <em>FOXP3</em>’ expression.</p>
<p>Below, we could rank the Treg cells by their smoothed <em>FOXP3</em> expression, as a way of measuring our confidence in the cell type call.</p>
<p>If we kept all of our Treg cells as Tregs, then 12.7% would be <em>FOXP3</em>+. If we keep only the top 100 cells, 42% would be <em>FOXP3</em>+.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>met <span class="ot">&lt;-</span> <span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"foxp3smooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"FOXP3"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"foxp3raw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"FOXP3"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="do">### calculate pct of actual FOXP3 positive Tregs, ranking by smoothed FOXP3</span></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>treg_filter <span class="ot">&lt;-</span> met[clust<span class="sc">==</span><span class="st">"Treg"</span>][<span class="fu">order</span>(<span class="sc">-</span>foxp3smooth)][,rnk<span class="sc">:</span><span class="er">=</span><span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>treg_filter[,pct_foxp3_positive<span class="sc">:</span><span class="er">=</span><span class="fu">cumsum</span>(foxp3raw <span class="sc">&gt;</span> <span class="dv">0</span>)<span class="sc">/</span>(rnk)]</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(treg_filter[<span class="fu">seq</span>(<span class="dv">100</span>,.N,<span class="at">length.out=</span><span class="dv">20</span>)]</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>       ,<span class="fu">aes</span>(<span class="at">x=</span>rnk, <span class="at">y=</span>pct_foxp3_positive)) <span class="sc">+</span> </span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>) <span class="sc">+</span> </span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">n.breaks=</span><span class="dv">20</span>) <span class="sc">+</span> </span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">n.breaks=</span><span class="dv">20</span>) <span class="sc">+</span> </span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>rnk, <span class="at">y=</span>pct_foxp3_positive <span class="sc">+</span> <span class="fl">0.01</span>, <span class="at">label=</span><span class="fu">formatC</span>(pct_foxp3_positive,<span class="dv">3</span>))) <span class="sc">+</span> </span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Proportion of FOXP3+ Treg cells when filtering by smoothed expression"</span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">x=</span><span class="st">"# of Treg cells kept"</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">y=</span><span class="st">"Proportion of FOXP3+ cells"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/filter_treg_bar.png" class="img-fluid figure-img" width="2254"></p>
</figure>
</div>
</div>
</div>
<p>We can also make a few scatter plots to show the bivariate distributions of some of these T cell markers, and color the cells by their original cell type labels. We can certainly see that each marker has a cluster of cells with strong, high-expression. On the other hand, there is a gray area at which unambiguous delineation between cell types becomes challenging.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd3gsmooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"CD3G"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd3graw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"CD3G"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd4smooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"CD4"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd4raw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"CD4"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd8bsmooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"CD8B"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd8braw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"CD8B"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>scatterp1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(met, <span class="fu">aes</span>(cd3gsmooth, foxp3smooth,<span class="at">color=</span>clust)) <span class="sc">+</span> </span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> ctpal, <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span> </span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>) <span class="sc">+</span> </span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a>scatterp1 <span class="ot">&lt;-</span> ggExtra<span class="sc">::</span><span class="fu">ggMarginal</span>(scatterp1, <span class="at">type =</span> <span class="st">"histogram"</span>)</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(scatterp1)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>scatterp2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(met, <span class="fu">aes</span>(cd3gsmooth, cd8bsmooth,<span class="at">color=</span>clust)) <span class="sc">+</span> </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> ctpal, <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span> </span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>) <span class="sc">+</span> </span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>scatterp2 <span class="ot">&lt;-</span> ggExtra<span class="sc">::</span><span class="fu">ggMarginal</span>(scatterp2, <span class="at">type =</span> <span class="st">"histogram"</span>)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(scatterp2)</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a>scatterp3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(met, <span class="fu">aes</span>(cd4smooth, foxp3smooth,<span class="at">color=</span>clust)) <span class="sc">+</span> </span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>) <span class="sc">+</span> </span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> ctpal, <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span> </span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>scatterp3 <span class="ot">&lt;-</span> ggExtra<span class="sc">::</span><span class="fu">ggMarginal</span>(scatterp3, <span class="at">type =</span> <span class="st">"histogram"</span>)</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(scatterp1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p><em>CD3G</em> vs.&nbsp;<em>FOXP3</em>:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/scatterp1.png" class="img-fluid figure-img" width="1838"></p>
</figure>
</div>
</div>
</div>
<p><em>CD3G</em> vs.&nbsp;<em>CD8B</em>:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/scatterp2.png" class="img-fluid figure-img" width="2399"></p>
</figure>
</div>
</div>
</div>
<p><em>CD4</em> vs.&nbsp;<em>FOXP3</em>:</p>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/scatterp3.png" class="img-fluid figure-img" width="1838"></p>
</figure>
</div>
</div>
</div>
<p>Here is a quick comparison of what the UMAP might look like after refining Tregs, using different filters.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ii <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>)){</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  met[,clust_filtered<span class="sc">:</span><span class="er">=</span>clust]</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  threshold <span class="ot">&lt;-</span> treg_filter[<span class="fu">seq</span>(<span class="dv">100</span>,.N,<span class="at">length.out=</span><span class="dv">20</span>)][ii][[<span class="st">"foxp3smooth"</span>]]</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  met[foxp3smooth <span class="sc">&lt;=</span>  threshold <span class="sc">&amp;</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>        clust_filtered<span class="sc">==</span><span class="st">"Treg"</span>,clust_filtered<span class="sc">:</span><span class="er">=</span><span class="st">"T CD4"</span>]</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>  sem<span class="sc">@</span>meta.data<span class="sc">$</span>clust_filtered <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>  sem <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">AddMetaData</span>(sem, <span class="at">metadata =</span> <span class="fu">data.frame</span>(met[,.(clust_filtered)], <span class="at">row.names =</span> met[[<span class="st">"cell_ID"</span>]]))</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(ii)]] <span class="ot">&lt;-</span> </span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"clust_filtered"</span>, sem</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">cls =</span> ctpal_sub, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">3</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="dv">5</span>)) <span class="sc">+</span> </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"Filtered Tregs (smooth FOXP3 &gt; "</span>,<span class="fu">formatC</span>(threshold,<span class="dv">3</span>), <span class="st">")"</span>)) <span class="sc">+</span> </span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plist,<span class="at">nrow=</span><span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/filter_treg_umaps.png" class="img-fluid figure-img" width="1096"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="other-marker-genes" class="level1" data-number="4">
<h1 data-number="4"><span class="header-section-number">4</span> Other marker genes</h1>
<p>It may be worth mentioning that marker genes for any cell type (not just T cells) could potentially be utilized for visualization or cell type refinement.</p>
<p>Here’s a few others that highlight key regions on the UMAP:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>mrks <span class="ot">&lt;-</span> <span class="fu">c</span>( </span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>  <span class="st">"FOXP3"</span> </span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CTLA4"</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"PDCD1"</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL32"</span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL7R"</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD8A"</span> </span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD8B"</span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD3E"</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD3D"</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD3G"</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"KLRK1"</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"MS4A1"</span></span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD19"</span></span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD79A"</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGKC"</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"JCHAIN"</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHG1"</span></span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHG2"</span></span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHM"</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHA1"</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"MZB1"</span></span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"XBP1"</span></span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CLEC10A"</span></span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL1B"</span></span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"ITGAX"</span></span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL3RA"</span></span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"GZMB"</span></span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"GNLY"</span></span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"GZMA"</span></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CXCL8"</span></span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD68"</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD163"</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"C1QA"</span>, <span class="st">"C1QB"</span>, <span class="st">"C1QC"</span>, <span class="st">"LYZ"</span>, <span class="st">"MMP9"</span>, <span class="st">"IL18"</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"VWF"</span>, <span class="st">"PECAM1"</span>, <span class="st">"RGS5"</span>, <span class="st">"SPARCL1"</span></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"KRT5"</span>, <span class="st">"KRT17"</span>, <span class="st">"KRT19"</span>, <span class="st">"EPCAM"</span></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"COL1A1"</span>, <span class="st">"COL1A2"</span>, <span class="st">"COL3A1"</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"TIGIT"</span></span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CPA3"</span>, <span class="st">"TPSB2"</span>, <span class="st">"TPSAB1"</span>, <span class="st">"KIT"</span></span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CSF3R"</span></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a>smoothed_markers <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data[mrks,] <span class="sc">%*%</span> smoother</span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(g <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"CD68"</span>, <span class="st">"GZMA"</span>, <span class="st">"GNLY"</span>, <span class="st">"MS4A1"</span>, <span class="st">"C1QC"</span>, <span class="st">"VWF"</span>)){</span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(g,<span class="st">".smooth"</span>)]] <span class="ot">&lt;-</span> </span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>    <span class="fu">marker_umap_plot</span>(smoothed_markers, g, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>)) <span class="sc">+</span> </span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"smooth "</span>,g))</span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist=</span>plist, <span class="at">nrow=</span><span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/smoothed_nontcell_markers.png" class="img-fluid figure-img" width="1710"></p>
</figure>
</div>
</div>
</div>
<p>A clustering analysis based only on smoothed expression of known marker genes could serve as a useful aid or even a standalone approach for cell typing. For example, here is a naive k-means clustering using a number of different marker genes, and with their clusters colored on the UMAP.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>kmd <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(Matrix<span class="sc">::</span><span class="fu">t</span>(smoothed_markers))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>kmd <span class="ot">&lt;-</span> kmd[,<span class="fu">lapply</span>(.SD, scale)]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>kmclus <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(kmd, <span class="at">centers=</span><span class="dv">20</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(kmclus<span class="sc">$</span>cluster) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(smoothed_markers)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>met <span class="ot">&lt;-</span> <span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>met[<span class="fu">match</span>(<span class="fu">names</span>(kmclus<span class="sc">$</span>cluster),cell_ID),kmclust<span class="sc">:</span><span class="er">=</span><span class="fu">as.character</span>(kmclus<span class="sc">$</span>cluster)]</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>sem<span class="sc">@</span>meta.data<span class="sc">$</span>kmclust <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">AddMetaData</span>(sem, <span class="at">metadata =</span> <span class="fu">data.frame</span>(met[,.(kmclust)], <span class="at">row.names=</span>met[[<span class="st">"cell_ID"</span>]]))</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a><span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"kmclust"</span>, sem)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="./figures/kmeans_smooth_markers.png" class="img-fluid figure-img" width="1710"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="conclusion" class="level1" data-number="5">
<h1 data-number="5"><span class="header-section-number">5</span> Conclusion</h1>
<p>In this vignette, I discussed a method that can be used to get an <strong>expected</strong> expression of a marker gene in a cell, given that cell’s nearest neighbors in UMAP space, along with potential applications to visualization and cell type refinement.</p>


<!-- -->


</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography" id="quarto-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list">
<div id="ref-Danaher2022" class="csl-entry" role="listitem">
Danaher, Patrick, Edward Zhao, Zhi Yang, David Ross, Mark Gregory, Zach Reitz, Tae K. Kim, et al. 2022. <span>“Insitutype: Likelihood-Based Cell Typing for Single Cell Spatial Transcriptomics.”</span> <em>bioRxiv</em>. <a href="https://doi.org/10.1101/2022.10.19.512902">https://doi.org/10.1101/2022.10.19.512902</a>.
</div>
<div id="ref-Lause2021" class="csl-entry" role="listitem">
Lause, Jan, Philipp Berens, and Dmitry Kobak. 2021. <span>“Analytic Pearson Residuals for Normalization of Single-Cell RNA-Seq UMI Data.”</span> <em>Genome Biology</em> 22 (September): 258. <a href="https://doi.org/10.1186/s13059-021-02451-7">https://doi.org/10.1186/s13059-021-02451-7</a>.
</div>
</div></section></div></main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleGiscusIfUsed = (isAlternate, darkModeDefault) => {
    const baseTheme = document.querySelector('#giscus-base-theme')?.value ?? 'light';
    const alternateTheme = document.querySelector('#giscus-alt-theme')?.value ?? 'dark';
    let newTheme = '';
    if(darkModeDefault) {
      newTheme = isAlternate ? baseTheme : alternateTheme;
    } else {
      newTheme = isAlternate ? alternateTheme : baseTheme;
    }
    const changeGiscusTheme = () => {
      // From: https://github.com/giscus/giscus/issues/336
      const sendMessage = (message) => {
        const iframe = document.querySelector('iframe.giscus-frame');
        if (!iframe) return;
        iframe.contentWindow.postMessage({ giscus: message }, 'https://giscus.app');
      }
      sendMessage({
        setConfig: {
          theme: newTheme
        }
      });
    }
    const isGiscussLoaded = window.document.querySelector('iframe.giscus-frame') !== null;
    if (isGiscussLoaded) {
      changeGiscusTheme();
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  const darkModeDefault = false;
  let localAlternateSentinel = darkModeDefault ? 'alternate' : 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
    toggleGiscusIfUsed(toAlternate, darkModeDefault);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    // For code content inside modals, clipBoardJS needs to be initialized with a container option
    // TODO: Check when it could be a function (https://github.com/zenorocha/clipboard.js/issues/860)
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/nanostring-biostats\.github\.io\/CosMx-Analysis-Scratch-Space");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb17" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "Applications for visualization and cell typing using 'smoothed' marker genes" </span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="co">  - name: Dan McGuire</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co">    orcid: 0009-0006-0286-9625</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co">    affiliations:</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="co">      - ref: nstg</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a><span class="co">      - ref: dan11mcguire</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="an">toc:</span><span class="co"> true</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="an">toc-title:</span><span class="co"> Contents</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="an">toc-depth:</span><span class="co"> 3</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a><span class="an">toc-expand:</span><span class="co"> 2</span></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="an">toc-location:</span><span class="co"> left</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a><span class="an">number-sections:</span><span class="co"> true</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="an">number-depth:</span><span class="co"> 4</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> "2024-06-19"</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a><span class="an">categories:</span><span class="co"> [visualization, cell typing]</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="an">image:</span><span class="co"> figures/smooth_foxp3_thumbnail.png</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a><span class="an">description:</span><span class="co"> "Applications for visualization and cell typing using 'smoothed' marker genes based on expression nearest neighbors"</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="an">draft:</span><span class="co"> false</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a><span class="an">format:</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a><span class="co">  html: </span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a><span class="co">    code-tools: true</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a><span class="fu"># Introduction </span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>Marker genes are genes that are expressed primarily within a single cell type, and are often used to delineate and label clusters during cell typing in a scRNAseq analysis.</span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>With spatially-resolved transcriptomics (SRT) data, a number of factors can contribute to challenges in a cell typing analysis, as well as hinder our ability to visualize our favorite marker genes. For example, these factors may include lower sensitivity compared to scRNAseq data, background due to autoflourescence, and segmentation error.</span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>A common feedback for new SRT analysts when dealing with SRT data may be along the lines of "Why don't I see gene 'X' in a majority of cells for cell type 'Y'?".  </span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a>In this post we'll discuss:</span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Why counts of a single marker gene are not definitive of cell type</span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How to derive more useful / less noisy / "smoothed" expression of marker genes</span>
<span id="cb17-38"><a href="#cb17-38" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>How to perform fine-grained subtyping using smoothed marker genes, with T-cell subtyping as motivation</span>
<span id="cb17-39"><a href="#cb17-39" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-40"><a href="#cb17-40" aria-hidden="true" tabindex="-1"></a><span class="fu"># A motivating example: Noisy T-cell typing in non-small cell lung cancer tissue</span></span>
<span id="cb17-41"><a href="#cb17-41" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-42"><a href="#cb17-42" aria-hidden="true" tabindex="-1"></a>For example, here is a data set below consisting of non-small cell lung cancer tissues.</span>
<span id="cb17-43"><a href="#cb17-43" aria-hidden="true" tabindex="-1"></a>Let's load it in, run some quick unrefined-cell typing using <span class="co">[</span><span class="ot">InSituType</span><span class="co">](https://github.com/Nanostring-Biostats/InSituType?tab=readme-ov-file)</span>{target="_blank"} <span class="co">[</span><span class="ot">@Danaher2022</span><span class="co">]</span>, and take a look at the initial results.</span>
<span id="cb17-44"><a href="#cb17-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-45"><a href="#cb17-45" aria-hidden="true" tabindex="-1"></a><span class="in">```{r load_libraries, message=FALSE, eval = FALSE}</span></span>
<span id="cb17-46"><a href="#cb17-46" aria-hidden="true" tabindex="-1"></a><span class="in">library(ggplot2)</span></span>
<span id="cb17-47"><a href="#cb17-47" aria-hidden="true" tabindex="-1"></a><span class="in">library(ggrepel)</span></span>
<span id="cb17-48"><a href="#cb17-48" aria-hidden="true" tabindex="-1"></a><span class="in">library(RColorBrewer)</span></span>
<span id="cb17-49"><a href="#cb17-49" aria-hidden="true" tabindex="-1"></a><span class="in">library(data.table)</span></span>
<span id="cb17-50"><a href="#cb17-50" aria-hidden="true" tabindex="-1"></a><span class="in">library(InSituType)</span></span>
<span id="cb17-51"><a href="#cb17-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-52"><a href="#cb17-52" aria-hidden="true" tabindex="-1"></a><span class="in">sem &lt;- readRDS("seurat_object.Rds")</span></span>
<span id="cb17-53"><a href="#cb17-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-54"><a href="#cb17-54" aria-hidden="true" tabindex="-1"></a><span class="in">### semi-supervised cell typing with 3 unsupervised clusters, using 'ioprofiles' reference matrix</span></span>
<span id="cb17-55"><a href="#cb17-55" aria-hidden="true" tabindex="-1"></a><span class="in">insitu &lt;- </span></span>
<span id="cb17-56"><a href="#cb17-56" aria-hidden="true" tabindex="-1"></a><span class="in">InSituType::insitutype(Matrix::t(sem[["RNA"]]@counts)</span></span>
<span id="cb17-57"><a href="#cb17-57" aria-hidden="true" tabindex="-1"></a><span class="in">                       ,neg = Matrix::colMeans(sem[["negprobes"]])</span></span>
<span id="cb17-58"><a href="#cb17-58" aria-hidden="true" tabindex="-1"></a><span class="in">                       ,reference_profiles = InSituType::ioprofiles</span></span>
<span id="cb17-59"><a href="#cb17-59" aria-hidden="true" tabindex="-1"></a><span class="in">                       ,n_clusts = 3</span></span>
<span id="cb17-60"><a href="#cb17-60" aria-hidden="true" tabindex="-1"></a><span class="in">                       )</span></span>
<span id="cb17-61"><a href="#cb17-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-62"><a href="#cb17-62" aria-hidden="true" tabindex="-1"></a><span class="in">### clean up cell type names and save results back to seurat object</span></span>
<span id="cb17-63"><a href="#cb17-63" aria-hidden="true" tabindex="-1"></a><span class="in">clustdt &lt;- data.table(clust_o = insitu$clust, row.names=names(insitu$clust))</span></span>
<span id="cb17-64"><a href="#cb17-64" aria-hidden="true" tabindex="-1"></a><span class="in">clustdt[,clust:=clust_o]</span></span>
<span id="cb17-65"><a href="#cb17-65" aria-hidden="true" tabindex="-1"></a><span class="in">clustdt[grep("T CD4", clust),clust:="T CD4"]</span></span>
<span id="cb17-66"><a href="#cb17-66" aria-hidden="true" tabindex="-1"></a><span class="in">clustdt[grep("T CD8", clust),clust:="T CD8"]</span></span>
<span id="cb17-67"><a href="#cb17-67" aria-hidden="true" tabindex="-1"></a><span class="in">clustdt[grep("B-cell", clust),clust:="B cell"]</span></span>
<span id="cb17-68"><a href="#cb17-68" aria-hidden="true" tabindex="-1"></a><span class="in">clustdt[,.N,by=.(clust)]</span></span>
<span id="cb17-69"><a href="#cb17-69" aria-hidden="true" tabindex="-1"></a><span class="in">sem@meta.data$clust &lt;- NULL</span></span>
<span id="cb17-70"><a href="#cb17-70" aria-hidden="true" tabindex="-1"></a><span class="in">sem &lt;- </span></span>
<span id="cb17-71"><a href="#cb17-71" aria-hidden="true" tabindex="-1"></a><span class="in">Seurat::AddMetaData(sem</span></span>
<span id="cb17-72"><a href="#cb17-72" aria-hidden="true" tabindex="-1"></a><span class="in">                    ,data.frame(clustdt[,.(clust, clust_o)]</span></span>
<span id="cb17-73"><a href="#cb17-73" aria-hidden="true" tabindex="-1"></a><span class="in">                                ,row.names=names(insitu$clust)</span></span>
<span id="cb17-74"><a href="#cb17-74" aria-hidden="true" tabindex="-1"></a><span class="in">                                )</span></span>
<span id="cb17-75"><a href="#cb17-75" aria-hidden="true" tabindex="-1"></a><span class="in">                    )</span></span>
<span id="cb17-76"><a href="#cb17-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-77"><a href="#cb17-77" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-78"><a href="#cb17-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-79"><a href="#cb17-79" aria-hidden="true" tabindex="-1"></a>Here we can plot the InSituType clusters ('clust') on the UMAP.</span>
<span id="cb17-80"><a href="#cb17-80" aria-hidden="true" tabindex="-1"></a>This UMAP was created using 'Analytic Pearson residuals' <span class="co">[</span><span class="ot">@Lause2021</span><span class="co">]</span>.</span>
<span id="cb17-81"><a href="#cb17-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-82"><a href="#cb17-82" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, umap_visualization_function, message=FALSE,eval=FALSE}</span></span>
<span id="cb17-83"><a href="#cb17-83" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb17-84"><a href="#cb17-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-85"><a href="#cb17-85" aria-hidden="true" tabindex="-1"></a><span class="in">umapf &lt;- function(umapreduc</span></span>
<span id="cb17-86"><a href="#cb17-86" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,clustercol</span></span>
<span id="cb17-87"><a href="#cb17-87" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,semuse</span></span>
<span id="cb17-88"><a href="#cb17-88" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,cls=NULL</span></span>
<span id="cb17-89"><a href="#cb17-89" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,xlim = NULL</span></span>
<span id="cb17-90"><a href="#cb17-90" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,ylim = NULL){</span></span>
<span id="cb17-91"><a href="#cb17-91" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-92"><a href="#cb17-92" aria-hidden="true" tabindex="-1"></a><span class="in">  umapd &lt;-  </span></span>
<span id="cb17-93"><a href="#cb17-93" aria-hidden="true" tabindex="-1"></a><span class="in">    data.table(semuse@reductions[[umapreduc]]@cell.embeddings</span></span>
<span id="cb17-94"><a href="#cb17-94" aria-hidden="true" tabindex="-1"></a><span class="in">               ,keep.rownames = TRUE)</span></span>
<span id="cb17-95"><a href="#cb17-95" aria-hidden="true" tabindex="-1"></a><span class="in">  setnames(umapd, c(names(umapd)[2:3]), c("UMAP_1", "UMAP_2"))</span></span>
<span id="cb17-96"><a href="#cb17-96" aria-hidden="true" tabindex="-1"></a><span class="in">  obsmrk &lt;- merge(data.table(semuse@meta.data), umapd</span></span>
<span id="cb17-97"><a href="#cb17-97" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,by.x="cell_ID"</span></span>
<span id="cb17-98"><a href="#cb17-98" aria-hidden="true" tabindex="-1"></a><span class="in">                  , by.y="rn")</span></span>
<span id="cb17-99"><a href="#cb17-99" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-100"><a href="#cb17-100" aria-hidden="true" tabindex="-1"></a><span class="in">  obstxt &lt;- obsmrk[,lapply(.SD, median),by=c(clustercol),.SDcols=paste0("UMAP_",1:2)]</span></span>
<span id="cb17-101"><a href="#cb17-101" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-102"><a href="#cb17-102" aria-hidden="true" tabindex="-1"></a><span class="in">    p &lt;- </span></span>
<span id="cb17-103"><a href="#cb17-103" aria-hidden="true" tabindex="-1"></a><span class="in">      ggplot(obsmrk, aes(UMAP_1, UMAP_2, color=.data[[clustercol]])) + </span></span>
<span id="cb17-104"><a href="#cb17-104" aria-hidden="true" tabindex="-1"></a><span class="in">      geom_point(size=0.2) + </span></span>
<span id="cb17-105"><a href="#cb17-105" aria-hidden="true" tabindex="-1"></a><span class="in">      theme_bw() + coord_fixed(xlim=xlim, ylim=ylim) + </span></span>
<span id="cb17-106"><a href="#cb17-106" aria-hidden="true" tabindex="-1"></a><span class="in">      geom_label_repel(data=obstxt, aes(x=UMAP_1, y=UMAP_2, label=.data[[clustercol]]),show.legend=FALSE</span></span>
<span id="cb17-107"><a href="#cb17-107" aria-hidden="true" tabindex="-1"></a><span class="in">                       ,inherit.aes=FALSE,color='black')</span></span>
<span id="cb17-108"><a href="#cb17-108" aria-hidden="true" tabindex="-1"></a><span class="in">    </span></span>
<span id="cb17-109"><a href="#cb17-109" aria-hidden="true" tabindex="-1"></a><span class="in">  if(is.null(cls)){</span></span>
<span id="cb17-110"><a href="#cb17-110" aria-hidden="true" tabindex="-1"></a><span class="in">    p &lt;- p +  </span></span>
<span id="cb17-111"><a href="#cb17-111" aria-hidden="true" tabindex="-1"></a><span class="in">      scale_color_manual(values=rep(unname(pals::alphabet()), 3)</span></span>
<span id="cb17-112"><a href="#cb17-112" aria-hidden="true" tabindex="-1"></a><span class="in">                         ,guide=guide_legend(override.aes=list(size=4)))</span></span>
<span id="cb17-113"><a href="#cb17-113" aria-hidden="true" tabindex="-1"></a><span class="in">    </span></span>
<span id="cb17-114"><a href="#cb17-114" aria-hidden="true" tabindex="-1"></a><span class="in">  } else {</span></span>
<span id="cb17-115"><a href="#cb17-115" aria-hidden="true" tabindex="-1"></a><span class="in">    p &lt;- p +  </span></span>
<span id="cb17-116"><a href="#cb17-116" aria-hidden="true" tabindex="-1"></a><span class="in">      scale_color_manual(values=cls</span></span>
<span id="cb17-117"><a href="#cb17-117" aria-hidden="true" tabindex="-1"></a><span class="in">                         ,guide=guide_legend(override.aes=list(size=4)))</span></span>
<span id="cb17-118"><a href="#cb17-118" aria-hidden="true" tabindex="-1"></a><span class="in">  }</span></span>
<span id="cb17-119"><a href="#cb17-119" aria-hidden="true" tabindex="-1"></a><span class="in">  return(p)</span></span>
<span id="cb17-120"><a href="#cb17-120" aria-hidden="true" tabindex="-1"></a><span class="in">}</span></span>
<span id="cb17-121"><a href="#cb17-121" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-122"><a href="#cb17-122" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-123"><a href="#cb17-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-124"><a href="#cb17-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-125"><a href="#cb17-125" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eval=FALSE, message=FALSE}</span></span>
<span id="cb17-126"><a href="#cb17-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-127"><a href="#cb17-127" aria-hidden="true" tabindex="-1"></a><span class="in">ctpal &lt;- c('#C20088','#005C31','#2BCE48'</span></span>
<span id="cb17-128"><a href="#cb17-128" aria-hidden="true" tabindex="-1"></a><span class="in">           ,'#4C005C','#F0A0FF','#003380'</span></span>
<span id="cb17-129"><a href="#cb17-129" aria-hidden="true" tabindex="-1"></a><span class="in">           ,'#FFCC99','#8F7C00','#9DCC00'</span></span>
<span id="cb17-130"><a href="#cb17-130" aria-hidden="true" tabindex="-1"></a><span class="in">           ,'#191919','#94FFB5','#0075DC'</span></span>
<span id="cb17-131"><a href="#cb17-131" aria-hidden="true" tabindex="-1"></a><span class="in">           ,'#FFA8BB','#FFA405','#993F00'</span></span>
<span id="cb17-132"><a href="#cb17-132" aria-hidden="true" tabindex="-1"></a><span class="in">           ,'#808080')</span></span>
<span id="cb17-133"><a href="#cb17-133" aria-hidden="true" tabindex="-1"></a><span class="in">names(ctpal) &lt;- c('endothelial','mDC','plasmablast'</span></span>
<span id="cb17-134"><a href="#cb17-134" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,'b','B cell','pDC'</span></span>
<span id="cb17-135"><a href="#cb17-135" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,'macrophage','a','mast'</span></span>
<span id="cb17-136"><a href="#cb17-136" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,'T CD4','fibroblast','T CD8'</span></span>
<span id="cb17-137"><a href="#cb17-137" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,'NK','Treg','neutrophil'</span></span>
<span id="cb17-138"><a href="#cb17-138" aria-hidden="true" tabindex="-1"></a><span class="in">                  ,'c')</span></span>
<span id="cb17-139"><a href="#cb17-139" aria-hidden="true" tabindex="-1"></a><span class="in">umapf("pearsonumap", "clust", sem, cls = ctpal)</span></span>
<span id="cb17-140"><a href="#cb17-140" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-141"><a href="#cb17-141" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-144"><a href="#cb17-144" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-145"><a href="#cb17-145" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-146"><a href="#cb17-146" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-147"><a href="#cb17-147" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-148"><a href="#cb17-148" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/umap_clust_all.png"</span>)</span>
<span id="cb17-149"><a href="#cb17-149" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-150"><a href="#cb17-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-151"><a href="#cb17-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-152"><a href="#cb17-152" aria-hidden="true" tabindex="-1"></a>Let's zoom in on some of the supervised clusters we hoped to identify below.</span>
<span id="cb17-153"><a href="#cb17-153" aria-hidden="true" tabindex="-1"></a>We can see that some of the major cell types are somewhat clearly delineated on the UMAP.</span>
<span id="cb17-154"><a href="#cb17-154" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-155"><a href="#cb17-155" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Lymphocytes; T-cell types (T CD8, T CD4, and Treg) are clustered together, and near the 'B cell' cluster.</span>
<span id="cb17-156"><a href="#cb17-156" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Myeloid cell types; (macrophage, pDC, mDC) are clustered in a similar area.</span>
<span id="cb17-157"><a href="#cb17-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-158"><a href="#cb17-158" aria-hidden="true" tabindex="-1"></a>But there does appear to potentially be some noise in delineating cell types within those major categories.</span>
<span id="cb17-159"><a href="#cb17-159" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eval=FALSE, message=FALSE}</span></span>
<span id="cb17-160"><a href="#cb17-160" aria-hidden="true" tabindex="-1"></a><span class="in">umapf("pearsonumap", "clust", sem, cls = ctpal, xlim = c(-5.5, 6.5), ylim = c(-0.01, 7.2))</span></span>
<span id="cb17-161"><a href="#cb17-161" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-162"><a href="#cb17-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-163"><a href="#cb17-163" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-166"><a href="#cb17-166" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-167"><a href="#cb17-167" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-168"><a href="#cb17-168" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-169"><a href="#cb17-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-170"><a href="#cb17-170" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/umap_clust_imm.png"</span>)</span>
<span id="cb17-171"><a href="#cb17-171" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-172"><a href="#cb17-172" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-173"><a href="#cb17-173" aria-hidden="true" tabindex="-1"></a>We can focus on T cells in this dataset as a driving example for using smoothing as an approach for addressing challenging cell typing and visualization problems. </span>
<span id="cb17-174"><a href="#cb17-174" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-175"><a href="#cb17-175" aria-hidden="true" tabindex="-1"></a>Canonical marker genes for T cells include</span>
<span id="cb17-176"><a href="#cb17-176" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-177"><a href="#cb17-177" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>*CD3* (expected to be expressed in all T-cell types), </span>
<span id="cb17-178"><a href="#cb17-178" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>*CD4* (commonly used together with *CD3* to identify T CD4 cells.  *CD4* can also be expressed in myeloid cells.)</span>
<span id="cb17-179"><a href="#cb17-179" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>*FOXP3* (commonly used together with *CD3* to identify Treg cells. Treg cells are a special subset of T CD4 cells)</span>
<span id="cb17-180"><a href="#cb17-180" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>*CD8A* and *CD8B* (used together with *CD3* to identify T CD8 cells)</span>
<span id="cb17-181"><a href="#cb17-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-182"><a href="#cb17-182" aria-hidden="true" tabindex="-1"></a>First, let's take a look at the relative frequency of our T-cell clusters 'Treg', 'T CD8', and 'T CD4'.</span>
<span id="cb17-183"><a href="#cb17-183" aria-hidden="true" tabindex="-1"></a>Tregs are supposed to be a rare sub type of T CD4 cells. Yet we have more than 3x as many Tregs called than we do T CD4.  </span>
<span id="cb17-184"><a href="#cb17-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-185"><a href="#cb17-185" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-186"><a href="#cb17-186" aria-hidden="true" tabindex="-1"></a><span class="in">```{r eval=FALSE, message=FALSE}</span></span>
<span id="cb17-187"><a href="#cb17-187" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb17-188"><a href="#cb17-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-189"><a href="#cb17-189" aria-hidden="true" tabindex="-1"></a><span class="in">bard &lt;- data.table(sem@meta.data)[,.N,by=.(clust)]</span></span>
<span id="cb17-190"><a href="#cb17-190" aria-hidden="true" tabindex="-1"></a><span class="in">bard[,clust:=factor(clust, levels=bard[order(-N),clust])]</span></span>
<span id="cb17-191"><a href="#cb17-191" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(bard[grep("^T",clust)], aes(x=clust, y=N,fill=clust)) + </span></span>
<span id="cb17-192"><a href="#cb17-192" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() + </span></span>
<span id="cb17-193"><a href="#cb17-193" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(text=element_text(size=16)) + </span></span>
<span id="cb17-194"><a href="#cb17-194" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_bar(stat='identity') +</span></span>
<span id="cb17-195"><a href="#cb17-195" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_fill_manual(values=ctpal, guide = guide_legend(reverse=TRUE)) + </span></span>
<span id="cb17-196"><a href="#cb17-196" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_y_continuous(n.breaks=12, labels = scales::comma, name = "# of cells") + </span></span>
<span id="cb17-197"><a href="#cb17-197" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title="# of cells called by T cell subtype") + </span></span>
<span id="cb17-198"><a href="#cb17-198" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(aes(x=clust, y=N + 200, label=scales::comma(N))) + </span></span>
<span id="cb17-199"><a href="#cb17-199" aria-hidden="true" tabindex="-1"></a><span class="in">  coord_flip()</span></span>
<span id="cb17-200"><a href="#cb17-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-201"><a href="#cb17-201" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-202"><a href="#cb17-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-205"><a href="#cb17-205" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-206"><a href="#cb17-206" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-207"><a href="#cb17-207" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-208"><a href="#cb17-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-209"><a href="#cb17-209" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/tcell_calls_barplot.png"</span>)</span>
<span id="cb17-210"><a href="#cb17-210" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-211"><a href="#cb17-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-212"><a href="#cb17-212" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-213"><a href="#cb17-213" aria-hidden="true" tabindex="-1"></a>Do we really have this many Treg cells in our data?</span>
<span id="cb17-214"><a href="#cb17-214" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-215"><a href="#cb17-215" aria-hidden="true" tabindex="-1"></a>To gather more evidence toward answering this kind of question, we can calculate the proportion of cells in each cell type expressing a particular marker gene, along with a fold change comparison for each cluster relative to other clusters.</span>
<span id="cb17-216"><a href="#cb17-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-217"><a href="#cb17-217" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE,message=FALSE}</span></span>
<span id="cb17-218"><a href="#cb17-218" aria-hidden="true" tabindex="-1"></a><span class="in">#| code-fold: true</span></span>
<span id="cb17-219"><a href="#cb17-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-220"><a href="#cb17-220" aria-hidden="true" tabindex="-1"></a><span class="in">totalcount_norm &lt;- function(sm){</span></span>
<span id="cb17-221"><a href="#cb17-221" aria-hidden="true" tabindex="-1"></a><span class="in">    libsizes &lt;- Matrix::colSums(sm)</span></span>
<span id="cb17-222"><a href="#cb17-222" aria-hidden="true" tabindex="-1"></a><span class="in">    scalefactor &lt;- mean(libsizes)</span></span>
<span id="cb17-223"><a href="#cb17-223" aria-hidden="true" tabindex="-1"></a><span class="in">    libsizes[libsizes==0] &lt;- 1</span></span>
<span id="cb17-224"><a href="#cb17-224" aria-hidden="true" tabindex="-1"></a><span class="in">    normed &lt;- sm %*% Matrix::Diagonal(x=scalefactor/libsizes)</span></span>
<span id="cb17-225"><a href="#cb17-225" aria-hidden="true" tabindex="-1"></a><span class="in">    dimnames(normed) &lt;- dimnames(sm)</span></span>
<span id="cb17-226"><a href="#cb17-226" aria-hidden="true" tabindex="-1"></a><span class="in">    return(normed)</span></span>
<span id="cb17-227"><a href="#cb17-227" aria-hidden="true" tabindex="-1"></a><span class="in">  }</span></span>
<span id="cb17-228"><a href="#cb17-228" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-229"><a href="#cb17-229" aria-hidden="true" tabindex="-1"></a><span class="in">clusterwise_fold_change_stats &lt;- function(cnts=NULL, normed = NULL, metainfo, clustercol){</span></span>
<span id="cb17-230"><a href="#cb17-230" aria-hidden="true" tabindex="-1"></a><span class="in">  if(missing(normed)){</span></span>
<span id="cb17-231"><a href="#cb17-231" aria-hidden="true" tabindex="-1"></a><span class="in">    normed &lt;- totalcount_norm(cnts) </span></span>
<span id="cb17-232"><a href="#cb17-232" aria-hidden="true" tabindex="-1"></a><span class="in">  }</span></span>
<span id="cb17-233"><a href="#cb17-233" aria-hidden="true" tabindex="-1"></a><span class="in">  outl &lt;- list()</span></span>
<span id="cb17-234"><a href="#cb17-234" aria-hidden="true" tabindex="-1"></a><span class="in">  for(ii in unique(metainfo[[clustercol]])){</span></span>
<span id="cb17-235"><a href="#cb17-235" aria-hidden="true" tabindex="-1"></a><span class="in">    cells_ii &lt;- metainfo[metainfo[[clustercol]]==ii,cell_ID]</span></span>
<span id="cb17-236"><a href="#cb17-236" aria-hidden="true" tabindex="-1"></a><span class="in">    cells_iiprime &lt;- metainfo[metainfo[[clustercol]]!=ii,cell_ID]</span></span>
<span id="cb17-237"><a href="#cb17-237" aria-hidden="true" tabindex="-1"></a><span class="in">    cluster_expr_ii  &lt;- Matrix::rowMeans(normed[,cells_ii,drop=FALSE])</span></span>
<span id="cb17-238"><a href="#cb17-238" aria-hidden="true" tabindex="-1"></a><span class="in">    cluster_expr_iiprime &lt;- Matrix::rowMeans(normed[,cells_iiprime,drop=FALSE]) </span></span>
<span id="cb17-239"><a href="#cb17-239" aria-hidden="true" tabindex="-1"></a><span class="in">    cluster_prop_ii &lt;- Matrix::rowMeans(normed[,cells_ii,drop=FALSE] &gt; 0) </span></span>
<span id="cb17-240"><a href="#cb17-240" aria-hidden="true" tabindex="-1"></a><span class="in">    cluster_prop_iiprime &lt;- Matrix::rowMeans(normed[,cells_iiprime,drop=FALSE] &gt; 0) </span></span>
<span id="cb17-241"><a href="#cb17-241" aria-hidden="true" tabindex="-1"></a><span class="in">    fctbl &lt;- data.table(cluster=ii</span></span>
<span id="cb17-242"><a href="#cb17-242" aria-hidden="true" tabindex="-1"></a><span class="in">                        ,cluster_expr = cluster_expr_ii</span></span>
<span id="cb17-243"><a href="#cb17-243" aria-hidden="true" tabindex="-1"></a><span class="in">                        ,clusterprime_expr = cluster_expr_iiprime</span></span>
<span id="cb17-244"><a href="#cb17-244" aria-hidden="true" tabindex="-1"></a><span class="in">                        ,gene = names(cluster_expr_ii)</span></span>
<span id="cb17-245"><a href="#cb17-245" aria-hidden="true" tabindex="-1"></a><span class="in">                        ,cluster_prop = cluster_prop_ii</span></span>
<span id="cb17-246"><a href="#cb17-246" aria-hidden="true" tabindex="-1"></a><span class="in">                        ,clusterprime_prop = cluster_prop_iiprime</span></span>
<span id="cb17-247"><a href="#cb17-247" aria-hidden="true" tabindex="-1"></a><span class="in">    )[,typ:=clustercol]</span></span>
<span id="cb17-248"><a href="#cb17-248" aria-hidden="true" tabindex="-1"></a><span class="in">    fctbl[,fold_change:=cluster_expr / clusterprime_expr]</span></span>
<span id="cb17-249"><a href="#cb17-249" aria-hidden="true" tabindex="-1"></a><span class="in">    fctbl[,fold_change_prop:=cluster_prop / clusterprime_prop]</span></span>
<span id="cb17-250"><a href="#cb17-250" aria-hidden="true" tabindex="-1"></a><span class="in">    outl[[paste0(ii)]] &lt;- copy(fctbl) </span></span>
<span id="cb17-251"><a href="#cb17-251" aria-hidden="true" tabindex="-1"></a><span class="in">  } </span></span>
<span id="cb17-252"><a href="#cb17-252" aria-hidden="true" tabindex="-1"></a><span class="in">  return(fc = rbindlist(outl))</span></span>
<span id="cb17-253"><a href="#cb17-253" aria-hidden="true" tabindex="-1"></a><span class="in">}</span></span>
<span id="cb17-254"><a href="#cb17-254" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-255"><a href="#cb17-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-256"><a href="#cb17-256" aria-hidden="true" tabindex="-1"></a>We can see that *FOXP3* is expressed in a much larger frequency in Treg cells relative to other T-cell subtypes-- this is good.  </span>
<span id="cb17-257"><a href="#cb17-257" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-258"><a href="#cb17-258" aria-hidden="true" tabindex="-1"></a>However, only 12.7% of our Treg cells are *FOXP3* positive, which may be a concern given that we have so many more Treg cells than other T-cell subtypes.</span>
<span id="cb17-259"><a href="#cb17-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-260"><a href="#cb17-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-261"><a href="#cb17-261" aria-hidden="true" tabindex="-1"></a><span class="in">```{r, eval=FALSE, message=FALSE}</span></span>
<span id="cb17-262"><a href="#cb17-262" aria-hidden="true" tabindex="-1"></a><span class="in">cluster_level_stats &lt;-  </span></span>
<span id="cb17-263"><a href="#cb17-263" aria-hidden="true" tabindex="-1"></a><span class="in">clusterwise_fold_change_stats(cnts = sem[["RNA"]]@counts</span></span>
<span id="cb17-264"><a href="#cb17-264" aria-hidden="true" tabindex="-1"></a><span class="in">                   ,metainfo = data.table(sem@meta.data)</span></span>
<span id="cb17-265"><a href="#cb17-265" aria-hidden="true" tabindex="-1"></a><span class="in">                   ,clustercol = "clust")</span></span>
<span id="cb17-266"><a href="#cb17-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-267"><a href="#cb17-267" aria-hidden="true" tabindex="-1"></a><span class="in">bard &lt;- cluster_level_stats[gene=="FOXP3"][grep("^T",cluster)]</span></span>
<span id="cb17-268"><a href="#cb17-268" aria-hidden="true" tabindex="-1"></a><span class="in">bard[,cluster:=factor(cluster, levels=bard[order(-cluster_prop),cluster])]</span></span>
<span id="cb17-269"><a href="#cb17-269" aria-hidden="true" tabindex="-1"></a><span class="in">ggplot(bard, aes(x=cluster, y=cluster_prop, fill=cluster)) + </span></span>
<span id="cb17-270"><a href="#cb17-270" aria-hidden="true" tabindex="-1"></a><span class="in">  theme_bw() + </span></span>
<span id="cb17-271"><a href="#cb17-271" aria-hidden="true" tabindex="-1"></a><span class="in">  theme(text=element_text(size=16)) + </span></span>
<span id="cb17-272"><a href="#cb17-272" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_bar(stat='identity') +</span></span>
<span id="cb17-273"><a href="#cb17-273" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_fill_manual(values=ctpal, guide = guide_legend(reverse=TRUE)) + </span></span>
<span id="cb17-274"><a href="#cb17-274" aria-hidden="true" tabindex="-1"></a><span class="in">  scale_y_continuous(n.breaks=12, labels = scales::comma, name = "Proportion of cells expressing the FOXP3 marker") + </span></span>
<span id="cb17-275"><a href="#cb17-275" aria-hidden="true" tabindex="-1"></a><span class="in">  labs(title="FOXP3") + </span></span>
<span id="cb17-276"><a href="#cb17-276" aria-hidden="true" tabindex="-1"></a><span class="in">  geom_text(aes(x=cluster, y=cluster_prop + 0.01, label=formatC(cluster_prop,3))) + </span></span>
<span id="cb17-277"><a href="#cb17-277" aria-hidden="true" tabindex="-1"></a><span class="in">  coord_flip()</span></span>
<span id="cb17-278"><a href="#cb17-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-279"><a href="#cb17-279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-280"><a href="#cb17-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-281"><a href="#cb17-281" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-284"><a href="#cb17-284" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-285"><a href="#cb17-285" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-286"><a href="#cb17-286" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-287"><a href="#cb17-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-288"><a href="#cb17-288" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/tcell_foxp3_pct_barplot.png"</span>)</span>
<span id="cb17-289"><a href="#cb17-289" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-290"><a href="#cb17-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-291"><a href="#cb17-291" aria-hidden="true" tabindex="-1"></a>At this stage, one may reasonably wonder: </span>
<span id="cb17-292"><a href="#cb17-292" aria-hidden="true" tabindex="-1"></a>"If we define a Treg cell based on *FOXP3* expression, why not call a cell  "Treg" **only** if it is expressing *FOXP3*?"</span>
<span id="cb17-293"><a href="#cb17-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-294"><a href="#cb17-294" aria-hidden="true" tabindex="-1"></a>Here we can illustrate the challenge this would present by plotting *FOXP3* expression on the UMAP.  On the one hand, if we look closely, we can see what looks like might be a "hot spot" on the UMAP where we are calling Treg cells, and we also have higher *FOXP3* expression.</span>
<span id="cb17-295"><a href="#cb17-295" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-296"><a href="#cb17-296" aria-hidden="true" tabindex="-1"></a>But setting a threshold based on *FOXP3* expression to define our Treg cells would be futile.  </span>
<span id="cb17-297"><a href="#cb17-297" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-298"><a href="#cb17-298" aria-hidden="true" tabindex="-1"></a>We have many cells near our '*FOXP3* hotspot' which may very well be Tregs, but do not have any counts of the marker gene.</span>
<span id="cb17-299"><a href="#cb17-299" aria-hidden="true" tabindex="-1"></a>On the other hand, we can see cells of all types, all over the UMAP, which have one or several counts of *FOXP3* expressed possibly due to some form of background.</span>
<span id="cb17-300"><a href="#cb17-300" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-303"><a href="#cb17-303" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-304"><a href="#cb17-304" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-305"><a href="#cb17-305" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-306"><a href="#cb17-306" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-307"><a href="#cb17-307" aria-hidden="true" tabindex="-1"></a>marker_umap_plot <span class="ot">&lt;-</span> <span class="cf">function</span>(marker_matrix, marker, sem</span>
<span id="cb17-308"><a href="#cb17-308" aria-hidden="true" tabindex="-1"></a>                             ,umapreduc, <span class="at">scale_expr =</span> <span class="cn">TRUE</span></span>
<span id="cb17-309"><a href="#cb17-309" aria-hidden="true" tabindex="-1"></a>                             ,<span class="at">xlim=</span><span class="cn">NULL</span>,<span class="at">ylim=</span><span class="cn">NULL</span>){</span>
<span id="cb17-310"><a href="#cb17-310" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-311"><a href="#cb17-311" aria-hidden="true" tabindex="-1"></a>  pd <span class="ot">&lt;-</span>  </span>
<span id="cb17-312"><a href="#cb17-312" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.table</span>(sem<span class="sc">@</span>reductions[[umapreduc]]<span class="sc">@</span>cell.embeddings</span>
<span id="cb17-313"><a href="#cb17-313" aria-hidden="true" tabindex="-1"></a>               ,<span class="at">keep.rownames =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-314"><a href="#cb17-314" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setnames</span>(pd, <span class="fu">c</span>(<span class="fu">names</span>(pd)[<span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>]), <span class="fu">c</span>(<span class="st">"cell_ID"</span>, <span class="st">"UMAP_1"</span>, <span class="st">"UMAP_2"</span>))</span>
<span id="cb17-315"><a href="#cb17-315" aria-hidden="true" tabindex="-1"></a>  pd[<span class="fu">match</span>(<span class="fu">colnames</span>(marker_matrix), cell_ID),(<span class="fu">c</span>(marker))<span class="sc">:</span><span class="er">=</span>marker_matrix[marker,]]</span>
<span id="cb17-316"><a href="#cb17-316" aria-hidden="true" tabindex="-1"></a>  <span class="fu">setkeyv</span>(pd, marker)</span>
<span id="cb17-317"><a href="#cb17-317" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(scale_expr){</span>
<span id="cb17-318"><a href="#cb17-318" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> </span>
<span id="cb17-319"><a href="#cb17-319" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggplot</span>(pd </span>
<span id="cb17-320"><a href="#cb17-320" aria-hidden="true" tabindex="-1"></a>             ,<span class="fu">aes</span>(<span class="at">x=</span>UMAP_1, <span class="at">y =</span> UMAP_2, <span class="at">color=</span><span class="fu">scale</span>(.data[[marker]]))) </span>
<span id="cb17-321"><a href="#cb17-321" aria-hidden="true" tabindex="-1"></a>      </span>
<span id="cb17-322"><a href="#cb17-322" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb17-323"><a href="#cb17-323" aria-hidden="true" tabindex="-1"></a>    p <span class="ot">&lt;-</span> </span>
<span id="cb17-324"><a href="#cb17-324" aria-hidden="true" tabindex="-1"></a>      <span class="fu">ggplot</span>(pd </span>
<span id="cb17-325"><a href="#cb17-325" aria-hidden="true" tabindex="-1"></a>             ,<span class="fu">aes</span>(<span class="at">x=</span>UMAP_1, <span class="at">y =</span> UMAP_2, <span class="at">color=</span>.data[[marker]])) </span>
<span id="cb17-326"><a href="#cb17-326" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb17-327"><a href="#cb17-327" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb17-328"><a href="#cb17-328" aria-hidden="true" tabindex="-1"></a>  p <span class="ot">&lt;-</span> p <span class="sc">+</span></span>
<span id="cb17-329"><a href="#cb17-329" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb17-330"><a href="#cb17-330" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_color_gradientn</span>(<span class="at">colors=</span><span class="fu">brewer.pal</span>(<span class="dv">9</span>, <span class="st">"Reds"</span>)) <span class="sc">+</span> </span>
<span id="cb17-331"><a href="#cb17-331" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb17-332"><a href="#cb17-332" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">n.breaks=</span><span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb17-333"><a href="#cb17-333" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">n.breaks=</span><span class="dv">10</span>) <span class="sc">+</span> </span>
<span id="cb17-334"><a href="#cb17-334" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span>marker) <span class="sc">+</span> <span class="fu">coord_fixed</span>(<span class="at">xlim=</span>xlim,<span class="at">ylim=</span>ylim) </span>
<span id="cb17-335"><a href="#cb17-335" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(p)</span>
<span id="cb17-336"><a href="#cb17-336" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-337"><a href="#cb17-337" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-338"><a href="#cb17-338" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-339"><a href="#cb17-339" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-342"><a href="#cb17-342" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-343"><a href="#cb17-343" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-344"><a href="#cb17-344" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-345"><a href="#cb17-345" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-346"><a href="#cb17-346" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-347"><a href="#cb17-347" aria-hidden="true" tabindex="-1"></a>raw_foxp3_plot <span class="ot">&lt;-</span> </span>
<span id="cb17-348"><a href="#cb17-348" aria-hidden="true" tabindex="-1"></a>  <span class="fu">marker_umap_plot</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts, <span class="st">"FOXP3"</span>, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb17-349"><a href="#cb17-349" aria-hidden="true" tabindex="-1"></a>                   ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>),<span class="at">scale=</span><span class="cn">FALSE</span>) <span class="sc">+</span> </span>
<span id="cb17-350"><a href="#cb17-350" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"raw FOXP3"</span>)</span>
<span id="cb17-351"><a href="#cb17-351" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-352"><a href="#cb17-352" aria-hidden="true" tabindex="-1"></a>ctpal_sub <span class="ot">&lt;-</span> ctpal</span>
<span id="cb17-353"><a href="#cb17-353" aria-hidden="true" tabindex="-1"></a>ctpal_sub[<span class="fu">grep</span>(<span class="st">"^T"</span>,<span class="fu">names</span>(ctpal_sub),<span class="at">invert=</span><span class="cn">TRUE</span>)] <span class="ot">&lt;-</span> <span class="st">"grey77"</span></span>
<span id="cb17-354"><a href="#cb17-354" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(</span>
<span id="cb17-355"><a href="#cb17-355" aria-hidden="true" tabindex="-1"></a>  raw_foxp3_plot</span>
<span id="cb17-356"><a href="#cb17-356" aria-hidden="true" tabindex="-1"></a>  ,<span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"clust"</span>, sem</span>
<span id="cb17-357"><a href="#cb17-357" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">cls =</span> ctpal_sub, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>))</span>
<span id="cb17-358"><a href="#cb17-358" aria-hidden="true" tabindex="-1"></a>  ,<span class="at">nrow=</span><span class="dv">2</span></span>
<span id="cb17-359"><a href="#cb17-359" aria-hidden="true" tabindex="-1"></a>  ) <span class="sc">+</span> </span>
<span id="cb17-360"><a href="#cb17-360" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb17-361"><a href="#cb17-361" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb17-362"><a href="#cb17-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-363"><a href="#cb17-363" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-364"><a href="#cb17-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-367"><a href="#cb17-367" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-368"><a href="#cb17-368" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-369"><a href="#cb17-369" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-370"><a href="#cb17-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-371"><a href="#cb17-371" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/foxp3raw.png"</span>)</span>
<span id="cb17-372"><a href="#cb17-372" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-373"><a href="#cb17-373" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-374"><a href="#cb17-374" aria-hidden="true" tabindex="-1"></a><span class="fu"># Using 'smoothed' FOXP3 to highlight Treg focal point and refine cell types.</span></span>
<span id="cb17-375"><a href="#cb17-375" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-376"><a href="#cb17-376" aria-hidden="true" tabindex="-1"></a>One potential approach to cleaning this up is to use some form of 'smoothed' marker gene expression, rather than the raw *FOXP3* counts.</span>
<span id="cb17-377"><a href="#cb17-377" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-378"><a href="#cb17-378" aria-hidden="true" tabindex="-1"></a>We can describe the calculation for 'smoothed' marker expression in two basic steps:</span>
<span id="cb17-379"><a href="#cb17-379" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-380"><a href="#cb17-380" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Find the $K$ nearest neighbors of each cell in UMAP space.</span>
<span id="cb17-381"><a href="#cb17-381" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Average the raw (or normalized) expression of the gene across that cells nearest neighbors.</span>
<span id="cb17-382"><a href="#cb17-382" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-383"><a href="#cb17-383" aria-hidden="true" tabindex="-1"></a>There's a few different ways we might think about motivating this approach.  The UMAP is an approximate manifold projection, which takes a low (2-d) representation of our high-dimensional (960 genes) data set.</span>
<span id="cb17-384"><a href="#cb17-384" aria-hidden="true" tabindex="-1"></a>The manifold is 'locally connected', meaning cells near each other in UMAP space are also similar to each other in expression space.</span>
<span id="cb17-385"><a href="#cb17-385" aria-hidden="true" tabindex="-1"></a>By averaging the expression of a marker gene of interest across cells with similar profiles, we can make a simple imputation of what we'd expect to see for the marker gene, given other cells of similar expression profiles.</span>
<span id="cb17-386"><a href="#cb17-386" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-387"><a href="#cb17-387" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-388"><a href="#cb17-388" aria-hidden="true" tabindex="-1"></a>This function below identifies the nearest neighbors in *UMAP* space, and makes a smoothing matrix, which can be used to get the average expression of any gene among a given cell's nearest neighbors.</span>
<span id="cb17-389"><a href="#cb17-389" aria-hidden="true" tabindex="-1"></a>Here is a simple implementation of the function, to take a 'Seurat' object as input, but could easily be modified for other data formats.</span>
<span id="cb17-390"><a href="#cb17-390" aria-hidden="true" tabindex="-1"></a><span class="in">```{r,eval=FALSE,message=FALSE}</span></span>
<span id="cb17-391"><a href="#cb17-391" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-392"><a href="#cb17-392" aria-hidden="true" tabindex="-1"></a><span class="in">## create a column-standardized nearest neighbor matrix </span></span>
<span id="cb17-393"><a href="#cb17-393" aria-hidden="true" tabindex="-1"></a><span class="in">## based on nearest neighbors in umap coordinates</span></span>
<span id="cb17-394"><a href="#cb17-394" aria-hidden="true" tabindex="-1"></a><span class="in">#' </span></span>
<span id="cb17-395"><a href="#cb17-395" aria-hidden="true" tabindex="-1"></a><span class="in">#' @param sem a seureat object, which contains the UMAP reduction.</span></span>
<span id="cb17-396"><a href="#cb17-396" aria-hidden="true" tabindex="-1"></a><span class="in">#' @param umapreduc name of the umap reduction, i.e. "umap"</span></span>
<span id="cb17-397"><a href="#cb17-397" aria-hidden="true" tabindex="-1"></a><span class="in">#' @param n_neighbor number of nearest neighbors to be identified.</span></span>
<span id="cb17-398"><a href="#cb17-398" aria-hidden="true" tabindex="-1"></a><span class="in">#' </span></span>
<span id="cb17-399"><a href="#cb17-399" aria-hidden="true" tabindex="-1"></a><span class="in">#' @return "smoother", a smoothing matrix. expression (genes x cells ) x  smoother(cells x cells) , </span></span>
<span id="cb17-400"><a href="#cb17-400" aria-hidden="true" tabindex="-1"></a><span class="in">#' will create an averaged expression across nearest neighbors.</span></span>
<span id="cb17-401"><a href="#cb17-401" aria-hidden="true" tabindex="-1"></a><span class="in">#' </span></span>
<span id="cb17-402"><a href="#cb17-402" aria-hidden="true" tabindex="-1"></a><span class="in">umap_nn &lt;- function(sem, umapreduc, n_neighbors=100){</span></span>
<span id="cb17-403"><a href="#cb17-403" aria-hidden="true" tabindex="-1"></a><span class="in"> </span></span>
<span id="cb17-404"><a href="#cb17-404" aria-hidden="true" tabindex="-1"></a><span class="in">  ## extract umap coordinates </span></span>
<span id="cb17-405"><a href="#cb17-405" aria-hidden="true" tabindex="-1"></a><span class="in">  umapd &lt;-  </span></span>
<span id="cb17-406"><a href="#cb17-406" aria-hidden="true" tabindex="-1"></a><span class="in">    data.table(sem@reductions[[umapreduc]]@cell.embeddings</span></span>
<span id="cb17-407"><a href="#cb17-407" aria-hidden="true" tabindex="-1"></a><span class="in">               ,keep.rownames = TRUE)</span></span>
<span id="cb17-408"><a href="#cb17-408" aria-hidden="true" tabindex="-1"></a><span class="in">  setnames(umapd, c(names(umapd)[1:3]), c("cell_ID", "UMAP_1", "UMAP_2"))</span></span>
<span id="cb17-409"><a href="#cb17-409" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-410"><a href="#cb17-410" aria-hidden="true" tabindex="-1"></a><span class="in">  ## identify nearest n_neighbors (+1 includes the cell as a neighbor to itself)</span></span>
<span id="cb17-411"><a href="#cb17-411" aria-hidden="true" tabindex="-1"></a><span class="in">  nn_umap &lt;- RANN::nn2(umapd[,.(UMAP_1, UMAP_2)],k = n_neighbors + 1)$nn.idx</span></span>
<span id="cb17-412"><a href="#cb17-412" aria-hidden="true" tabindex="-1"></a><span class="in">  nn_umap &lt;- data.table::melt(cbind(umapd[,.(cell_ID)], data.table(nn_umap))</span></span>
<span id="cb17-413"><a href="#cb17-413" aria-hidden="true" tabindex="-1"></a><span class="in">                              , id.vars=c("cell_ID", "V1"))</span></span>
<span id="cb17-414"><a href="#cb17-414" aria-hidden="true" tabindex="-1"></a><span class="in">  colnames(nn_umap) &lt;- c("cell_ID1", "cell_ID1_idx", "neighbor", "cell_ID2_idx")</span></span>
<span id="cb17-415"><a href="#cb17-415" aria-hidden="true" tabindex="-1"></a><span class="in">  nn_umap &lt;- merge(nn_umap</span></span>
<span id="cb17-416"><a href="#cb17-416" aria-hidden="true" tabindex="-1"></a><span class="in">                   , nn_umap[,.(cell_ID2=cell_ID1, cell_ID2_idx=cell_ID1_idx)][</span></span>
<span id="cb17-417"><a href="#cb17-417" aria-hidden="true" tabindex="-1"></a><span class="in">                     ,unique(.SD)],by="cell_ID2_idx")</span></span>
<span id="cb17-418"><a href="#cb17-418" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-419"><a href="#cb17-419" aria-hidden="true" tabindex="-1"></a><span class="in">  ## Cell x cell neighbor indicator matrix</span></span>
<span id="cb17-420"><a href="#cb17-420" aria-hidden="true" tabindex="-1"></a><span class="in">  wumap &lt;- Matrix::sparseMatrix(i = c(unique(nn_umap$cell_ID1_idx), nn_umap$cell_ID2_idx)</span></span>
<span id="cb17-421"><a href="#cb17-421" aria-hidden="true" tabindex="-1"></a><span class="in">                                ,j=c(unique(nn_umap$cell_ID1_idx), nn_umap$cell_ID1_idx)</span></span>
<span id="cb17-422"><a href="#cb17-422" aria-hidden="true" tabindex="-1"></a><span class="in">                                ,x=1)</span></span>
<span id="cb17-423"><a href="#cb17-423" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-424"><a href="#cb17-424" aria-hidden="true" tabindex="-1"></a><span class="in">  rownames(wumap) &lt;- colnames(wumap) &lt;- nn_umap[order(nn_umap$cell_ID1_idx),unique(cell_ID1)]</span></span>
<span id="cb17-425"><a href="#cb17-425" aria-hidden="true" tabindex="-1"></a><span class="in">  </span></span>
<span id="cb17-426"><a href="#cb17-426" aria-hidden="true" tabindex="-1"></a><span class="in">  ## Column standardize, so that columns (cells) sum to 1, and each neighbor given equal weight.</span></span>
<span id="cb17-427"><a href="#cb17-427" aria-hidden="true" tabindex="-1"></a><span class="in">  mumap &lt;- Matrix::sparseMatrix(i=1:ncol(wumap)</span></span>
<span id="cb17-428"><a href="#cb17-428" aria-hidden="true" tabindex="-1"></a><span class="in">                                ,j=1:ncol(wumap)</span></span>
<span id="cb17-429"><a href="#cb17-429" aria-hidden="true" tabindex="-1"></a><span class="in">                                ,x=1/Matrix::colSums(wumap)</span></span>
<span id="cb17-430"><a href="#cb17-430" aria-hidden="true" tabindex="-1"></a><span class="in">                                )</span></span>
<span id="cb17-431"><a href="#cb17-431" aria-hidden="true" tabindex="-1"></a><span class="in">  dimnames(mumap) &lt;- dimnames(wumap)</span></span>
<span id="cb17-432"><a href="#cb17-432" aria-hidden="true" tabindex="-1"></a><span class="in">  smoother &lt;- mumap %*% wumap</span></span>
<span id="cb17-433"><a href="#cb17-433" aria-hidden="true" tabindex="-1"></a><span class="in">  smoother &lt;- smoother[,colnames(sem)]</span></span>
<span id="cb17-434"><a href="#cb17-434" aria-hidden="true" tabindex="-1"></a><span class="in">  return(smoother)</span></span>
<span id="cb17-435"><a href="#cb17-435" aria-hidden="true" tabindex="-1"></a><span class="in">}</span></span>
<span id="cb17-436"><a href="#cb17-436" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-437"><a href="#cb17-437" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-438"><a href="#cb17-438" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-439"><a href="#cb17-439" aria-hidden="true" tabindex="-1"></a>Here, we can apply our 'smoothing matrix' to some T-cell marker genes and compare the difference on the UMAP.</span>
<span id="cb17-440"><a href="#cb17-440" aria-hidden="true" tabindex="-1"></a>Whereas raw marker expression is noisy and sparse, smoothed expression clearly highlights the focal points for each marker on the UMAP.</span>
<span id="cb17-441"><a href="#cb17-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-444"><a href="#cb17-444" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-445"><a href="#cb17-445" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-446"><a href="#cb17-446" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-447"><a href="#cb17-447" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-448"><a href="#cb17-448" aria-hidden="true" tabindex="-1"></a>smoother <span class="ot">&lt;-</span> <span class="fu">umap_nn</span>(sem, <span class="st">"pearsonumap"</span>, <span class="at">n_neighbors =</span> <span class="dv">500</span>)</span>
<span id="cb17-449"><a href="#cb17-449" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> <span class="fu">SetAssayData</span>(sem, <span class="st">"data"</span>, <span class="at">assay =</span> <span class="st">"RNA"</span>, <span class="at">new.data =</span> <span class="fu">totalcount_norm</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts))</span>
<span id="cb17-450"><a href="#cb17-450" aria-hidden="true" tabindex="-1"></a>smoothed_tcell_markers <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data[<span class="fu">c</span>(<span class="st">"FOXP3"</span>,<span class="st">"CD3G"</span>,<span class="st">"CD8B"</span>, <span class="st">"CD4"</span>),] <span class="sc">%*%</span> smoother</span>
<span id="cb17-451"><a href="#cb17-451" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-452"><a href="#cb17-452" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb17-453"><a href="#cb17-453" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(g <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"FOXP3"</span>, <span class="st">"CD8B"</span>,<span class="st">"CD3G"</span>)){</span>
<span id="cb17-454"><a href="#cb17-454" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(g,<span class="st">".smooth"</span>)]] <span class="ot">&lt;-</span> </span>
<span id="cb17-455"><a href="#cb17-455" aria-hidden="true" tabindex="-1"></a>    <span class="fu">marker_umap_plot</span>(smoothed_tcell_markers, g, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb17-456"><a href="#cb17-456" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>)) <span class="sc">+</span> </span>
<span id="cb17-457"><a href="#cb17-457" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"smooth "</span>,g))</span>
<span id="cb17-458"><a href="#cb17-458" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(g,<span class="st">".raw"</span>)]] <span class="ot">&lt;-</span> </span>
<span id="cb17-459"><a href="#cb17-459" aria-hidden="true" tabindex="-1"></a>    <span class="fu">marker_umap_plot</span>(sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts, g, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb17-460"><a href="#cb17-460" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>)) <span class="sc">+</span> </span>
<span id="cb17-461"><a href="#cb17-461" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"raw "</span>,g))</span>
<span id="cb17-462"><a href="#cb17-462" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-463"><a href="#cb17-463" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist=</span>plist, <span class="at">nrow=</span><span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb17-464"><a href="#cb17-464" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb17-465"><a href="#cb17-465" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb17-466"><a href="#cb17-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-467"><a href="#cb17-467" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-468"><a href="#cb17-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-471"><a href="#cb17-471" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-472"><a href="#cb17-472" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-473"><a href="#cb17-473" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-474"><a href="#cb17-474" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-475"><a href="#cb17-475" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/smooth_vs_raw.png"</span>)</span>
<span id="cb17-476"><a href="#cb17-476" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-477"><a href="#cb17-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-478"><a href="#cb17-478" aria-hidden="true" tabindex="-1"></a>From here, we could take a number of different approaches to 'clean up' our T-cell calls.</span>
<span id="cb17-479"><a href="#cb17-479" aria-hidden="true" tabindex="-1"></a>For example, because Tregs are a subset of T CD4 cells, we might consider reassigning them to T CD4 if they have low 'smoothed *FOXP3*' expression.</span>
<span id="cb17-480"><a href="#cb17-480" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-481"><a href="#cb17-481" aria-hidden="true" tabindex="-1"></a>Below, we could rank the Treg cells by their smoothed *FOXP3* expression, as a way of measuring our confidence in the cell type call.</span>
<span id="cb17-482"><a href="#cb17-482" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-483"><a href="#cb17-483" aria-hidden="true" tabindex="-1"></a>If we kept all of our Treg cells as Tregs, then 12.7% would be *FOXP3*+.</span>
<span id="cb17-484"><a href="#cb17-484" aria-hidden="true" tabindex="-1"></a>If we keep only the top 100 cells, 42% would be *FOXP3*+.  </span>
<span id="cb17-485"><a href="#cb17-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-488"><a href="#cb17-488" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-489"><a href="#cb17-489" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-490"><a href="#cb17-490" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-491"><a href="#cb17-491" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-492"><a href="#cb17-492" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-493"><a href="#cb17-493" aria-hidden="true" tabindex="-1"></a>met <span class="ot">&lt;-</span> <span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)</span>
<span id="cb17-494"><a href="#cb17-494" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"foxp3smooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"FOXP3"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-495"><a href="#cb17-495" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"foxp3raw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"FOXP3"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-496"><a href="#cb17-496" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-497"><a href="#cb17-497" aria-hidden="true" tabindex="-1"></a><span class="do">### calculate pct of actual FOXP3 positive Tregs, ranking by smoothed FOXP3</span></span>
<span id="cb17-498"><a href="#cb17-498" aria-hidden="true" tabindex="-1"></a>treg_filter <span class="ot">&lt;-</span> met[clust<span class="sc">==</span><span class="st">"Treg"</span>][<span class="fu">order</span>(<span class="sc">-</span>foxp3smooth)][,rnk<span class="sc">:</span><span class="er">=</span><span class="dv">1</span><span class="sc">:</span>.N]</span>
<span id="cb17-499"><a href="#cb17-499" aria-hidden="true" tabindex="-1"></a>treg_filter[,pct_foxp3_positive<span class="sc">:</span><span class="er">=</span><span class="fu">cumsum</span>(foxp3raw <span class="sc">&gt;</span> <span class="dv">0</span>)<span class="sc">/</span>(rnk)]</span>
<span id="cb17-500"><a href="#cb17-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-501"><a href="#cb17-501" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(treg_filter[<span class="fu">seq</span>(<span class="dv">100</span>,.N,<span class="at">length.out=</span><span class="dv">20</span>)]</span>
<span id="cb17-502"><a href="#cb17-502" aria-hidden="true" tabindex="-1"></a>       ,<span class="fu">aes</span>(<span class="at">x=</span>rnk, <span class="at">y=</span>pct_foxp3_positive)) <span class="sc">+</span> </span>
<span id="cb17-503"><a href="#cb17-503" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">stat=</span><span class="st">'identity'</span>) <span class="sc">+</span> </span>
<span id="cb17-504"><a href="#cb17-504" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb17-505"><a href="#cb17-505" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_y_continuous</span>(<span class="at">n.breaks=</span><span class="dv">20</span>) <span class="sc">+</span> </span>
<span id="cb17-506"><a href="#cb17-506" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">n.breaks=</span><span class="dv">20</span>) <span class="sc">+</span> </span>
<span id="cb17-507"><a href="#cb17-507" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_text</span>(<span class="fu">aes</span>(<span class="at">x=</span>rnk, <span class="at">y=</span>pct_foxp3_positive <span class="sc">+</span> <span class="fl">0.01</span>, <span class="at">label=</span><span class="fu">formatC</span>(pct_foxp3_positive,<span class="dv">3</span>))) <span class="sc">+</span> </span>
<span id="cb17-508"><a href="#cb17-508" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title=</span><span class="st">"Proportion of FOXP3+ Treg cells when filtering by smoothed expression"</span></span>
<span id="cb17-509"><a href="#cb17-509" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">x=</span><span class="st">"# of Treg cells kept"</span></span>
<span id="cb17-510"><a href="#cb17-510" aria-hidden="true" tabindex="-1"></a>       ,<span class="at">y=</span><span class="st">"Proportion of FOXP3+ cells"</span>)</span>
<span id="cb17-511"><a href="#cb17-511" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-512"><a href="#cb17-512" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-513"><a href="#cb17-513" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-514"><a href="#cb17-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-517"><a href="#cb17-517" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-518"><a href="#cb17-518" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-519"><a href="#cb17-519" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-520"><a href="#cb17-520" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-521"><a href="#cb17-521" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/filter_treg_bar.png"</span>)</span>
<span id="cb17-522"><a href="#cb17-522" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-523"><a href="#cb17-523" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-524"><a href="#cb17-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-525"><a href="#cb17-525" aria-hidden="true" tabindex="-1"></a>We can also make a few scatter plots to show the bivariate distributions of some of these T cell markers, and color the cells by their original cell type labels.</span>
<span id="cb17-526"><a href="#cb17-526" aria-hidden="true" tabindex="-1"></a>We can certainly see that each marker has a cluster of cells with strong, high-expression. On the other hand, there is a gray area at which unambiguous delineation between cell types becomes challenging.</span>
<span id="cb17-527"><a href="#cb17-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-530"><a href="#cb17-530" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-531"><a href="#cb17-531" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-532"><a href="#cb17-532" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb17-533"><a href="#cb17-533" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-534"><a href="#cb17-534" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-535"><a href="#cb17-535" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd3gsmooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"CD3G"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-536"><a href="#cb17-536" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd3graw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"CD3G"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-537"><a href="#cb17-537" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd4smooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"CD4"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-538"><a href="#cb17-538" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd4raw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"CD4"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-539"><a href="#cb17-539" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd8bsmooth"</span>]] <span class="ot">&lt;-</span> smoothed_tcell_markers[<span class="st">"CD8B"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-540"><a href="#cb17-540" aria-hidden="true" tabindex="-1"></a>met[[<span class="st">"cd8braw"</span>]] <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>counts[<span class="st">"CD8B"</span>,met[[<span class="st">"cell_ID"</span>]]]</span>
<span id="cb17-541"><a href="#cb17-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-542"><a href="#cb17-542" aria-hidden="true" tabindex="-1"></a>scatterp1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(met, <span class="fu">aes</span>(cd3gsmooth, foxp3smooth,<span class="at">color=</span>clust)) <span class="sc">+</span> </span>
<span id="cb17-543"><a href="#cb17-543" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb17-544"><a href="#cb17-544" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb17-545"><a href="#cb17-545" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> ctpal, <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span> </span>
<span id="cb17-546"><a href="#cb17-546" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>) <span class="sc">+</span> </span>
<span id="cb17-547"><a href="#cb17-547" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb17-548"><a href="#cb17-548" aria-hidden="true" tabindex="-1"></a>scatterp1 <span class="ot">&lt;-</span> ggExtra<span class="sc">::</span><span class="fu">ggMarginal</span>(scatterp1, <span class="at">type =</span> <span class="st">"histogram"</span>)</span>
<span id="cb17-549"><a href="#cb17-549" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(scatterp1)</span>
<span id="cb17-550"><a href="#cb17-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-551"><a href="#cb17-551" aria-hidden="true" tabindex="-1"></a>scatterp2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(met, <span class="fu">aes</span>(cd3gsmooth, cd8bsmooth,<span class="at">color=</span>clust)) <span class="sc">+</span> </span>
<span id="cb17-552"><a href="#cb17-552" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb17-553"><a href="#cb17-553" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb17-554"><a href="#cb17-554" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> ctpal, <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span> </span>
<span id="cb17-555"><a href="#cb17-555" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>) <span class="sc">+</span> </span>
<span id="cb17-556"><a href="#cb17-556" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb17-557"><a href="#cb17-557" aria-hidden="true" tabindex="-1"></a>scatterp2 <span class="ot">&lt;-</span> ggExtra<span class="sc">::</span><span class="fu">ggMarginal</span>(scatterp2, <span class="at">type =</span> <span class="st">"histogram"</span>)</span>
<span id="cb17-558"><a href="#cb17-558" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(scatterp2)</span>
<span id="cb17-559"><a href="#cb17-559" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-560"><a href="#cb17-560" aria-hidden="true" tabindex="-1"></a>scatterp3 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(met, <span class="fu">aes</span>(cd4smooth, foxp3smooth,<span class="at">color=</span>clust)) <span class="sc">+</span> </span>
<span id="cb17-561"><a href="#cb17-561" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">size=</span><span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb17-562"><a href="#cb17-562" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() <span class="sc">+</span> </span>
<span id="cb17-563"><a href="#cb17-563" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"left"</span>) <span class="sc">+</span> </span>
<span id="cb17-564"><a href="#cb17-564" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> ctpal, <span class="at">guide=</span><span class="fu">guide_legend</span>(<span class="at">override.aes=</span><span class="fu">list</span>(<span class="at">size=</span><span class="dv">4</span>))) <span class="sc">+</span> </span>
<span id="cb17-565"><a href="#cb17-565" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_fixed</span>()</span>
<span id="cb17-566"><a href="#cb17-566" aria-hidden="true" tabindex="-1"></a>scatterp3 <span class="ot">&lt;-</span> ggExtra<span class="sc">::</span><span class="fu">ggMarginal</span>(scatterp3, <span class="at">type =</span> <span class="st">"histogram"</span>)</span>
<span id="cb17-567"><a href="#cb17-567" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(scatterp1)</span>
<span id="cb17-568"><a href="#cb17-568" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-569"><a href="#cb17-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-570"><a href="#cb17-570" aria-hidden="true" tabindex="-1"></a>*CD3G* vs. *FOXP3*:</span>
<span id="cb17-573"><a href="#cb17-573" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-574"><a href="#cb17-574" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-575"><a href="#cb17-575" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-576"><a href="#cb17-576" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-577"><a href="#cb17-577" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/scatterp1.png"</span>)</span>
<span id="cb17-578"><a href="#cb17-578" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-579"><a href="#cb17-579" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-580"><a href="#cb17-580" aria-hidden="true" tabindex="-1"></a>*CD3G* vs. *CD8B*:</span>
<span id="cb17-583"><a href="#cb17-583" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-584"><a href="#cb17-584" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-585"><a href="#cb17-585" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-586"><a href="#cb17-586" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-587"><a href="#cb17-587" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/scatterp2.png"</span>)</span>
<span id="cb17-588"><a href="#cb17-588" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-589"><a href="#cb17-589" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-590"><a href="#cb17-590" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-591"><a href="#cb17-591" aria-hidden="true" tabindex="-1"></a>*CD4* vs. *FOXP3*:</span>
<span id="cb17-594"><a href="#cb17-594" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-595"><a href="#cb17-595" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-596"><a href="#cb17-596" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-597"><a href="#cb17-597" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-598"><a href="#cb17-598" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/scatterp3.png"</span>)</span>
<span id="cb17-599"><a href="#cb17-599" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-600"><a href="#cb17-600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-601"><a href="#cb17-601" aria-hidden="true" tabindex="-1"></a>Here is a quick comparison of what the UMAP might look like after refining Tregs, using different filters.</span>
<span id="cb17-602"><a href="#cb17-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-605"><a href="#cb17-605" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-606"><a href="#cb17-606" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-607"><a href="#cb17-607" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-608"><a href="#cb17-608" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-609"><a href="#cb17-609" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-610"><a href="#cb17-610" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb17-611"><a href="#cb17-611" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(ii <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>)){</span>
<span id="cb17-612"><a href="#cb17-612" aria-hidden="true" tabindex="-1"></a>  met[,clust_filtered<span class="sc">:</span><span class="er">=</span>clust]</span>
<span id="cb17-613"><a href="#cb17-613" aria-hidden="true" tabindex="-1"></a>  threshold <span class="ot">&lt;-</span> treg_filter[<span class="fu">seq</span>(<span class="dv">100</span>,.N,<span class="at">length.out=</span><span class="dv">20</span>)][ii][[<span class="st">"foxp3smooth"</span>]]</span>
<span id="cb17-614"><a href="#cb17-614" aria-hidden="true" tabindex="-1"></a>  met[foxp3smooth <span class="sc">&lt;=</span>  threshold <span class="sc">&amp;</span></span>
<span id="cb17-615"><a href="#cb17-615" aria-hidden="true" tabindex="-1"></a>        clust_filtered<span class="sc">==</span><span class="st">"Treg"</span>,clust_filtered<span class="sc">:</span><span class="er">=</span><span class="st">"T CD4"</span>]</span>
<span id="cb17-616"><a href="#cb17-616" aria-hidden="true" tabindex="-1"></a>  sem<span class="sc">@</span>meta.data<span class="sc">$</span>clust_filtered <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-617"><a href="#cb17-617" aria-hidden="true" tabindex="-1"></a>  sem <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">AddMetaData</span>(sem, <span class="at">metadata =</span> <span class="fu">data.frame</span>(met[,.(clust_filtered)], <span class="at">row.names =</span> met[[<span class="st">"cell_ID"</span>]]))</span>
<span id="cb17-618"><a href="#cb17-618" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(ii)]] <span class="ot">&lt;-</span> </span>
<span id="cb17-619"><a href="#cb17-619" aria-hidden="true" tabindex="-1"></a>  <span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"clust_filtered"</span>, sem</span>
<span id="cb17-620"><a href="#cb17-620" aria-hidden="true" tabindex="-1"></a>         ,<span class="at">cls =</span> ctpal_sub, <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1</span>, <span class="dv">3</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="fl">2.5</span>, <span class="dv">5</span>)) <span class="sc">+</span> </span>
<span id="cb17-621"><a href="#cb17-621" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"Filtered Tregs (smooth FOXP3 &gt; "</span>,<span class="fu">formatC</span>(threshold,<span class="dv">3</span>), <span class="st">")"</span>)) <span class="sc">+</span> </span>
<span id="cb17-622"><a href="#cb17-622" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">legend.position=</span><span class="st">"none"</span>)</span>
<span id="cb17-623"><a href="#cb17-623" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-624"><a href="#cb17-624" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist =</span> plist,<span class="at">nrow=</span><span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb17-625"><a href="#cb17-625" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb17-626"><a href="#cb17-626" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb17-627"><a href="#cb17-627" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-628"><a href="#cb17-628" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-629"><a href="#cb17-629" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-630"><a href="#cb17-630" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-633"><a href="#cb17-633" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-634"><a href="#cb17-634" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-635"><a href="#cb17-635" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-636"><a href="#cb17-636" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-637"><a href="#cb17-637" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/filter_treg_umaps.png"</span>)</span>
<span id="cb17-638"><a href="#cb17-638" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-639"><a href="#cb17-639" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-640"><a href="#cb17-640" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-641"><a href="#cb17-641" aria-hidden="true" tabindex="-1"></a><span class="fu"># Other marker genes</span></span>
<span id="cb17-642"><a href="#cb17-642" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-643"><a href="#cb17-643" aria-hidden="true" tabindex="-1"></a>It may be worth mentioning that marker genes for any cell type (not just T cells) could potentially be utilized for visualization or cell type refinement.</span>
<span id="cb17-644"><a href="#cb17-644" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-645"><a href="#cb17-645" aria-hidden="true" tabindex="-1"></a>Here's a few others that highlight key regions on the UMAP:</span>
<span id="cb17-646"><a href="#cb17-646" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-649"><a href="#cb17-649" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-650"><a href="#cb17-650" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-651"><a href="#cb17-651" aria-hidden="true" tabindex="-1"></a><span class="co">#| code-fold: true</span></span>
<span id="cb17-652"><a href="#cb17-652" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-653"><a href="#cb17-653" aria-hidden="true" tabindex="-1"></a>mrks <span class="ot">&lt;-</span> <span class="fu">c</span>( </span>
<span id="cb17-654"><a href="#cb17-654" aria-hidden="true" tabindex="-1"></a>  <span class="st">"FOXP3"</span> </span>
<span id="cb17-655"><a href="#cb17-655" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CTLA4"</span></span>
<span id="cb17-656"><a href="#cb17-656" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"PDCD1"</span></span>
<span id="cb17-657"><a href="#cb17-657" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL32"</span></span>
<span id="cb17-658"><a href="#cb17-658" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL7R"</span></span>
<span id="cb17-659"><a href="#cb17-659" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD8A"</span> </span>
<span id="cb17-660"><a href="#cb17-660" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD8B"</span></span>
<span id="cb17-661"><a href="#cb17-661" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD3E"</span></span>
<span id="cb17-662"><a href="#cb17-662" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD3D"</span></span>
<span id="cb17-663"><a href="#cb17-663" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD3G"</span></span>
<span id="cb17-664"><a href="#cb17-664" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"KLRK1"</span></span>
<span id="cb17-665"><a href="#cb17-665" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"MS4A1"</span></span>
<span id="cb17-666"><a href="#cb17-666" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD19"</span></span>
<span id="cb17-667"><a href="#cb17-667" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD79A"</span></span>
<span id="cb17-668"><a href="#cb17-668" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGKC"</span></span>
<span id="cb17-669"><a href="#cb17-669" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"JCHAIN"</span></span>
<span id="cb17-670"><a href="#cb17-670" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHG1"</span></span>
<span id="cb17-671"><a href="#cb17-671" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHG2"</span></span>
<span id="cb17-672"><a href="#cb17-672" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHM"</span></span>
<span id="cb17-673"><a href="#cb17-673" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IGHA1"</span></span>
<span id="cb17-674"><a href="#cb17-674" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"MZB1"</span></span>
<span id="cb17-675"><a href="#cb17-675" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"XBP1"</span></span>
<span id="cb17-676"><a href="#cb17-676" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CLEC10A"</span></span>
<span id="cb17-677"><a href="#cb17-677" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL1B"</span></span>
<span id="cb17-678"><a href="#cb17-678" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"ITGAX"</span></span>
<span id="cb17-679"><a href="#cb17-679" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"IL3RA"</span></span>
<span id="cb17-680"><a href="#cb17-680" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"GZMB"</span></span>
<span id="cb17-681"><a href="#cb17-681" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"GNLY"</span></span>
<span id="cb17-682"><a href="#cb17-682" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"GZMA"</span></span>
<span id="cb17-683"><a href="#cb17-683" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CXCL8"</span></span>
<span id="cb17-684"><a href="#cb17-684" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD68"</span></span>
<span id="cb17-685"><a href="#cb17-685" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CD163"</span></span>
<span id="cb17-686"><a href="#cb17-686" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"C1QA"</span>, <span class="st">"C1QB"</span>, <span class="st">"C1QC"</span>, <span class="st">"LYZ"</span>, <span class="st">"MMP9"</span>, <span class="st">"IL18"</span></span>
<span id="cb17-687"><a href="#cb17-687" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"VWF"</span>, <span class="st">"PECAM1"</span>, <span class="st">"RGS5"</span>, <span class="st">"SPARCL1"</span></span>
<span id="cb17-688"><a href="#cb17-688" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"KRT5"</span>, <span class="st">"KRT17"</span>, <span class="st">"KRT19"</span>, <span class="st">"EPCAM"</span></span>
<span id="cb17-689"><a href="#cb17-689" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"COL1A1"</span>, <span class="st">"COL1A2"</span>, <span class="st">"COL3A1"</span></span>
<span id="cb17-690"><a href="#cb17-690" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"TIGIT"</span></span>
<span id="cb17-691"><a href="#cb17-691" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CPA3"</span>, <span class="st">"TPSB2"</span>, <span class="st">"TPSAB1"</span>, <span class="st">"KIT"</span></span>
<span id="cb17-692"><a href="#cb17-692" aria-hidden="true" tabindex="-1"></a>  ,<span class="st">"CSF3R"</span></span>
<span id="cb17-693"><a href="#cb17-693" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb17-694"><a href="#cb17-694" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-695"><a href="#cb17-695" aria-hidden="true" tabindex="-1"></a>smoothed_markers <span class="ot">&lt;-</span> sem[[<span class="st">"RNA"</span>]]<span class="sc">@</span>data[mrks,] <span class="sc">%*%</span> smoother</span>
<span id="cb17-696"><a href="#cb17-696" aria-hidden="true" tabindex="-1"></a>plist <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb17-697"><a href="#cb17-697" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(g <span class="cf">in</span> <span class="fu">c</span>(<span class="st">"CD68"</span>, <span class="st">"GZMA"</span>, <span class="st">"GNLY"</span>, <span class="st">"MS4A1"</span>, <span class="st">"C1QC"</span>, <span class="st">"VWF"</span>)){</span>
<span id="cb17-698"><a href="#cb17-698" aria-hidden="true" tabindex="-1"></a>  plist[[<span class="fu">paste0</span>(g,<span class="st">".smooth"</span>)]] <span class="ot">&lt;-</span> </span>
<span id="cb17-699"><a href="#cb17-699" aria-hidden="true" tabindex="-1"></a>    <span class="fu">marker_umap_plot</span>(smoothed_markers, g, sem, <span class="st">"pearsonumap"</span></span>
<span id="cb17-700"><a href="#cb17-700" aria-hidden="true" tabindex="-1"></a>                     ,<span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">5.5</span>, <span class="fl">6.5</span>), <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">0.01</span>, <span class="fl">7.2</span>)) <span class="sc">+</span> </span>
<span id="cb17-701"><a href="#cb17-701" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">title=</span><span class="fu">paste0</span>(<span class="st">"smooth "</span>,g))</span>
<span id="cb17-702"><a href="#cb17-702" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-703"><a href="#cb17-703" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb17-704"><a href="#cb17-704" aria-hidden="true" tabindex="-1"></a>cowplot<span class="sc">::</span><span class="fu">plot_grid</span>(<span class="at">plotlist=</span>plist, <span class="at">nrow=</span><span class="dv">3</span>) <span class="sc">+</span> </span>
<span id="cb17-705"><a href="#cb17-705" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">plot.background =</span> <span class="fu">element_rect</span>(<span class="at">fill =</span> <span class="st">'white'</span>,<span class="at">color=</span><span class="st">'white'</span>)) <span class="sc">+</span> </span>
<span id="cb17-706"><a href="#cb17-706" aria-hidden="true" tabindex="-1"></a>  cowplot<span class="sc">::</span><span class="fu">panel_border</span>(<span class="at">remove=</span><span class="cn">TRUE</span>) </span>
<span id="cb17-707"><a href="#cb17-707" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-708"><a href="#cb17-708" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-709"><a href="#cb17-709" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-710"><a href="#cb17-710" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-713"><a href="#cb17-713" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-714"><a href="#cb17-714" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-715"><a href="#cb17-715" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-716"><a href="#cb17-716" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-717"><a href="#cb17-717" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/smoothed_nontcell_markers.png"</span>)</span>
<span id="cb17-718"><a href="#cb17-718" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-719"><a href="#cb17-719" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-720"><a href="#cb17-720" aria-hidden="true" tabindex="-1"></a>A clustering analysis based only on smoothed expression of known marker genes could serve as a useful aid or even a standalone approach for cell typing.</span>
<span id="cb17-721"><a href="#cb17-721" aria-hidden="true" tabindex="-1"></a>For example, here is a naive k-means clustering using a number of different marker genes, and with their clusters colored on the UMAP.</span>
<span id="cb17-722"><a href="#cb17-722" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-725"><a href="#cb17-725" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-726"><a href="#cb17-726" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb17-727"><a href="#cb17-727" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-728"><a href="#cb17-728" aria-hidden="true" tabindex="-1"></a>kmd <span class="ot">&lt;-</span> <span class="fu">as.data.table</span>(Matrix<span class="sc">::</span><span class="fu">t</span>(smoothed_markers))</span>
<span id="cb17-729"><a href="#cb17-729" aria-hidden="true" tabindex="-1"></a>kmd <span class="ot">&lt;-</span> kmd[,<span class="fu">lapply</span>(.SD, scale)]</span>
<span id="cb17-730"><a href="#cb17-730" aria-hidden="true" tabindex="-1"></a>kmclus <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(kmd, <span class="at">centers=</span><span class="dv">20</span>)</span>
<span id="cb17-731"><a href="#cb17-731" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(kmclus<span class="sc">$</span>cluster) <span class="ot">&lt;-</span> <span class="fu">colnames</span>(smoothed_markers)</span>
<span id="cb17-732"><a href="#cb17-732" aria-hidden="true" tabindex="-1"></a>met <span class="ot">&lt;-</span> <span class="fu">data.table</span>(sem<span class="sc">@</span>meta.data)</span>
<span id="cb17-733"><a href="#cb17-733" aria-hidden="true" tabindex="-1"></a>met[<span class="fu">match</span>(<span class="fu">names</span>(kmclus<span class="sc">$</span>cluster),cell_ID),kmclust<span class="sc">:</span><span class="er">=</span><span class="fu">as.character</span>(kmclus<span class="sc">$</span>cluster)]</span>
<span id="cb17-734"><a href="#cb17-734" aria-hidden="true" tabindex="-1"></a>sem<span class="sc">@</span>meta.data<span class="sc">$</span>kmclust <span class="ot">&lt;-</span> <span class="cn">NULL</span></span>
<span id="cb17-735"><a href="#cb17-735" aria-hidden="true" tabindex="-1"></a>sem <span class="ot">&lt;-</span> Seurat<span class="sc">::</span><span class="fu">AddMetaData</span>(sem, <span class="at">metadata =</span> <span class="fu">data.frame</span>(met[,.(kmclust)], <span class="at">row.names=</span>met[[<span class="st">"cell_ID"</span>]]))</span>
<span id="cb17-736"><a href="#cb17-736" aria-hidden="true" tabindex="-1"></a><span class="fu">umapf</span>(<span class="st">"pearsonumap"</span>, <span class="st">"kmclust"</span>, sem)</span>
<span id="cb17-737"><a href="#cb17-737" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-738"><a href="#cb17-738" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-739"><a href="#cb17-739" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-740"><a href="#cb17-740" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-743"><a href="#cb17-743" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb17-744"><a href="#cb17-744" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: true</span></span>
<span id="cb17-745"><a href="#cb17-745" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb17-746"><a href="#cb17-746" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-747"><a href="#cb17-747" aria-hidden="true" tabindex="-1"></a>knitr<span class="sc">::</span><span class="fu">include_graphics</span>(<span class="st">"./figures/kmeans_smooth_markers.png"</span>)</span>
<span id="cb17-748"><a href="#cb17-748" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb17-749"><a href="#cb17-749" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-750"><a href="#cb17-750" aria-hidden="true" tabindex="-1"></a><span class="fu"># Conclusion</span></span>
<span id="cb17-751"><a href="#cb17-751" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-752"><a href="#cb17-752" aria-hidden="true" tabindex="-1"></a>In this vignette, I discussed a method that can be used to get an **expected** expression of a marker gene in a cell, given that cell's nearest neighbors in UMAP space, along with potential applications to </span>
<span id="cb17-753"><a href="#cb17-753" aria-hidden="true" tabindex="-1"></a>visualization and cell type refinement.  </span>
<span id="cb17-754"><a href="#cb17-754" aria-hidden="true" tabindex="-1"></a></span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">
      <ul class="footer-items list-unstyled">
    <li class="nav-item">
    <a class="nav-link" href="../../license.html">
<p>License</p>
</a>
  </li>  
</ul>
    <div class="cookie-consent-footer"><a href="#" id="open_preferences_center">Cookie Preferences</a></div></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>




</body></html>